{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17bffff9-5c89-4285-be6f-23a1a8d2189c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "### Generate directories \n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "mkdir $HOME_DIRECTORY/scripts\n",
    "\n",
    "mkdir $SCRATCH_DIRECTORY/job_output\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/prepare_ref/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/trimming/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/bwa_mem/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/samtools_sort/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/samtools_view/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/MarkDuplicates/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/AddOrReplaceReadGroups/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/samtools_index/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/HaplotypeCaller/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/GenomicsDBImport/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/GenotypeGVCFs/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/bcftools_annotate/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/DP_range/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/SelectVariants_SNP/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/bcftools_split/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/bcftools_DP/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/VariantFiltration/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/callability\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/samtools_merge/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/accuMUlate/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/de_novo_mutation/\n",
    "\n",
    "mkdir $SCRATCH_DIRECTORY/ref_fasta/\n",
    "mkdir $SCRATCH_DIRECTORY/fastq_files/\n",
    "mkdir $SCRATCH_DIRECTORY/trimmed_seq/\n",
    "mkdir $SCRATCH_DIRECTORY/bam_files/\n",
    "mkdir $HOME_DIRECTORY/vcf_files/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/callability/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/callability/T1/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/callability/T2/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/callability/T5/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/GenomicsDBImport_database/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/GenomicsDBImport_temp/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/GenomicsDBImport_temp/T1_temp\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/GenomicsDBImport_temp/T2_temp\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/GenomicsDBImport_temp/T5_temp\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/GenotypeGVCFs/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/bcftools_annotate/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/SelectVariants_SNP/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/bcftools_DP/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/bcftools_DP/T1/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/bcftools_DP/T2/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/bcftools_DP/T5/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/VariantFiltration/\n",
    "mkdir $HOME_DIRECTORY/de_novo_mutation/\n",
    "mkdir $HOME_DIRECTORY/accuMUlate/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d97dff75-c17a-46f9-8163-49d10376f6f0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "### Define variables \n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "echo \"QF-1499-T1-10_S14_L008_R1_001.fastq.gz QF-1499-T1-10_S14_L008_R2_001.fastq.gz T1-10\n",
    "QF-1499-T1-19_S1_L001_R1_001.fastq.gz QF-1499-T1-19_S1_L001_R2_001.fastq.gz T1-19\n",
    "QF-1499-T1-37_S2_L001_R1_001.fastq.gz QF-1499-T1-37_S2_L001_R2_001.fastq.gz T1-37\n",
    "QF-1499-T1-46_S15_L008_R1_001.fastq.gz QF-1499-T1-46_S15_L008_R2_001.fastq.gz T1-46\n",
    "QF-1499-T1-57_S3_L001_R1_001.fastq.gz QF-1499-T1-57_S3_L001_R2_001.fastq.gz T1-57\n",
    "QF-1499-T1-6_S17_L008_R1_001.fastq.gz QF-1499-T1-6_S17_L008_R2_001.fastq.gz T1-6\n",
    "QF-1499-T2-1_S4_L001_R1_001.fastq.gz QF-1499-T2-1_S4_L001_R2_001.fastq.gz T2-1\n",
    "QF-1499-T2-22_S6_L001_R1_001.fastq.gz QF-1499-T2-22_S6_L001_R2_001.fastq.gz T2-22\n",
    "QF-1499-T2-34_S12_L008_R1_001.fastq.gz QF-1499-T2-34_S12_L008_R2_001.fastq.gz T2-34\n",
    "QF-1499-T2-4_S5_L001_R1_001.fastq.gz QF-1499-T2-4_S5_L001_R2_001.fastq.gz T2-4\n",
    "QF-1499-T2-42_S11_L008_R1_001.fastq.gz QF-1499-T2-42_S11_L008_R2_001.fastq.gz T2-42\n",
    "QF-1499-T2-52_S7_L001_R1_001.fastq.gz QF-1499-T2-52_S7_L001_R2_001.fastq.gz T2-52\n",
    "QF-1499-T5-45_S10_L008_R1_001.fastq.gz QF-1499-T5-45_S10_L008_R2_001.fastq.gz T5-45\n",
    "QF-1499-T5-5_S16_L008_R1_001.fastq.gz QF-1499-T5-5_S16_L008_R2_001.fastq.gz T5-5\n",
    "QF-1499-T5-70_S18_L008_R1_001.fastq.gz QF-1499-T5-70_S18_L008_R2_001.fastq.gz T5-70\n",
    "QF-1499-T5-81_S13_L008_R1_001.fastq.gz QF-1499-T5-81_S13_L008_R2_001.fastq.gz T5-81\n",
    "QF-1499-T5-86_S8_L001_R1_001.fastq.gz QF-1499-T5-86_S8_L001_R2_001.fastq.gz T5-86\n",
    "QF-1499-T5-87_S9_L001_R1_001.fastq.gz QF-1499-T5-87_S9_L001_R2_001.fastq.gz T5-87\"> $HOME_DIRECTORY/fastq_list\n",
    "\n",
    "\n",
    "echo \"T1\n",
    "T2\n",
    "T5\" > $HOME_DIRECTORY/treatment_list\n",
    "\n",
    "echo \"T1-10\n",
    "T1-19\n",
    "T1-37\n",
    "T1-46\n",
    "T1-57\n",
    "T1-6\" > $HOME_DIRECTORY/T1_sample_list\n",
    "\n",
    "echo \"$HOME_DIRECTORY/vcf_files/T1-10.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T1-19.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T1-37.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T1-46.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T1-57.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T1-6.g.vcf.gz\" > $HOME_DIRECTORY/T1_vcf_list\n",
    "\n",
    "paste -d\"\\t\" $HOME_DIRECTORY/T1_sample_list $HOME_DIRECTORY/T1_vcf_list > $HOME_DIRECTORY/T1_sample_map\n",
    "\n",
    "echo \"T2-1\n",
    "T2-22\n",
    "T2-34\n",
    "T2-4\n",
    "T2-42\n",
    "T2-52\" > $HOME_DIRECTORY/T2_sample_list\n",
    "\n",
    "echo \"$HOME_DIRECTORY/vcf_files/T2-1.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T2-22.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T2-34.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T2-4.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T2-42.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T2-52.g.vcf.gz\" > $HOME_DIRECTORY/T2_vcf_list\n",
    "\n",
    "paste -d\"\\t\" $HOME_DIRECTORY/T2_sample_list $HOME_DIRECTORY/T2_vcf_list > $HOME_DIRECTORY/T2_sample_map\n",
    "\n",
    "echo \"T5-45\n",
    "T5-5\n",
    "T5-70\n",
    "T5-81\n",
    "T5-86\n",
    "T5-87\" > $HOME_DIRECTORY/T5_sample_list\n",
    "\n",
    "echo \"$HOME_DIRECTORY/vcf_files/T5-45.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T5-5.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T5-70.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T5-81.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T5-86.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T5-87.g.vcf.gz\" > $HOME_DIRECTORY/T5_vcf_list\n",
    "\n",
    "paste -d\"\\t\" $HOME_DIRECTORY/T5_sample_list $HOME_DIRECTORY/T5_vcf_list > $HOME_DIRECTORY/T5_sample_map\n",
    "\n",
    "\n",
    "printf \"T1\\tT1-10\\t10\n",
    "T1\\tT1-19\\t11\n",
    "T1\\tT1-37\\t12\n",
    "T1\\tT1-46\\t13\n",
    "T1\\tT1-57\\t14\n",
    "T1\\tT1-6\\t15\n",
    "T2\\tT2-1\\t10\n",
    "T2\\tT2-22\\t11\n",
    "T2\\tT2-34\\t12\n",
    "T2\\tT2-4\\t13\n",
    "T2\\tT2-42\\t14\n",
    "T2\\tT2-52\\t15\n",
    "T5\\tT5-45\\t10\n",
    "T5\\tT5-5\\t11\n",
    "T5\\tT5-70\\t12\n",
    "T5\\tT5-81\\t13\n",
    "T5\\tT5-86\\t14\n",
    "T5\\tT5-87\\t15\" > $HOME_DIRECTORY/treatment_sample_field_for_DP\n",
    "\n",
    "printf \"T1 T1-10 0\n",
    "T1 T1-19 1\n",
    "T1 T1-37 2\n",
    "T1 T1-46 3\n",
    "T1 T1-57 4\n",
    "T1 T1-6 5\n",
    "T2 T2-1 0\n",
    "T2 T2-22 1\n",
    "T2 T2-34 2\n",
    "T2 T2-4 3\n",
    "T2 T2-42 4\n",
    "T2 T2-52 5\n",
    "T5 T5-45 0\n",
    "T5 T5-5 1\n",
    "T5 T5-70 2\n",
    "T5 T5-81 3\n",
    "T5 T5-86 4\n",
    "T5 T5-87 5\" > $HOME_DIRECTORY/treatment_sample_field_for_callability\n",
    "\n",
    "\n",
    "printf \"T1 T1-10 T1-19 T1-37 T1-46 T1-57 T1-6\n",
    "T2 T2-1 T2-22 T2-34 T2-4 T2-42 T2-52\n",
    "T5 T5-45 T5-5 T5-70 T5-81 T5-86 T5-87\" > $HOME_DIRECTORY/treatment_sample_wide\n",
    "\n",
    "echo \"T1 1 2.00 60.00 3.00 40.00 -12.50 -8.00\n",
    "T1 2 2.00 20.00 3.00 40.00 -12.50 -8.00\n",
    "T1 3 2.00 60.00 2.00 40.00 -12.50 -8.00\n",
    "T1 4 2.00 60.00 2.00 50.00 -12.50 -8.00\n",
    "T1 5 2.00 60.00 3.00 40.00 -6.00 -8.00\n",
    "T1 6 2.00 60.00 3.00 40.00 -12.50 -4.00\n",
    "T2 1 2.00 60.00 3.00 40.00 -12.50 -8.00\n",
    "T2 2 2.00 20.00 3.00 40.00 -12.50 -8.00\n",
    "T2 3 2.00 60.00 2.00 40.00 -12.50 -8.00\n",
    "T2 4 2.00 60.00 2.00 50.00 -12.50 -8.00\n",
    "T2 5 2.00 60.00 3.00 40.00 -6.00 -8.00\n",
    "T2 6 2.00 60.00 3.00 40.00 -12.50 -4.00\n",
    "T5 1 2.00 60.00 3.00 40.00 -12.50 -8.00\n",
    "T5 2 2.00 20.00 3.00 40.00 -12.50 -8.00\n",
    "T5 3 2.00 60.00 2.00 40.00 -12.50 -8.00\n",
    "T5 4 2.00 60.00 2.00 50.00 -12.50 -8.00\n",
    "T5 5 2.00 60.00 3.00 40.00 -6.00 -8.00\n",
    "T5 6 2.00 60.00 3.00 40.00 -12.50 -4.00\" > $HOME_DIRECTORY/VariantFiltration_parameter_combinations\n",
    "\n",
    "echo \"theta=0.0001\n",
    "mu=0.00000001\n",
    "nfreqs=0.25     0.25    0.25    0.25\n",
    "seq-error=0.01\n",
    "ploidy-ancestor=2\n",
    "ploidy-descendant=2\n",
    "phi-diploid=0.001\n",
    "phi-haploid=0.001\n",
    "\" > $HOME_DIRECTORY/accuMUlate_parameters\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e4d7b80-5334-47c0-a615-d688f0038e63",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Generate a symbolic link for each sample\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T1-19/QF-1499-T1-19_S1_L001_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-19_S1_L001_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T1-19/QF-1499-T1-19_S1_L001_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-19_S1_L001_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T1-37/QF-1499-T1-37_S2_L001_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-37_S2_L001_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T1-37/QF-1499-T1-37_S2_L001_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-37_S2_L001_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T1-57/QF-1499-T1-57_S3_L001_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-57_S3_L001_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T1-57/QF-1499-T1-57_S3_L001_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-57_S3_L001_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T2-1/QF-1499-T2-1_S4_L001_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-1_S4_L001_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T2-1/QF-1499-T2-1_S4_L001_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-1_S4_L001_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T2-22/QF-1499-T2-22_S6_L001_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-22_S6_L001_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T2-22/QF-1499-T2-22_S6_L001_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-22_S6_L001_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T2-4/QF-1499-T2-4_S5_L001_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-4_S5_L001_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T2-4/QF-1499-T2-4_S5_L001_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-4_S5_L001_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T2-52/QF-1499-T2-52_S7_L001_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-52_S7_L001_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T2-52/QF-1499-T2-52_S7_L001_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-52_S7_L001_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T5-86/QF-1499-T5-86_S8_L001_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-86_S8_L001_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T5-86/QF-1499-T5-86_S8_L001_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-86_S8_L001_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T5-87/QF-1499-T5-87_S9_L001_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-87_S9_L001_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T5-87/QF-1499-T5-87_S9_L001_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-87_S9_L001_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T1-10/QF-1499-T1-10_S14_L008_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-10_S14_L008_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T1-10/QF-1499-T1-10_S14_L008_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-10_S14_L008_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T1-46/QF-1499-T1-46_S15_L008_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-46_S15_L008_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T1-46/QF-1499-T1-46_S15_L008_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-46_S15_L008_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T1-6/QF-1499-T1-6_S17_L008_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-6_S17_L008_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T1-6/QF-1499-T1-6_S17_L008_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-6_S17_L008_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T2-34/QF-1499-T2-34_S12_L008_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-34_S12_L008_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T2-34/QF-1499-T2-34_S12_L008_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-34_S12_L008_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T2-42/QF-1499-T2-42_S11_L008_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-42_S11_L008_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T2-42/QF-1499-T2-42_S11_L008_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-42_S11_L008_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T5-45/QF-1499-T5-45_S10_L008_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-45_S10_L008_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T5-45/QF-1499-T5-45_S10_L008_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-45_S10_L008_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T5-5/QF-1499-T5-5_S16_L008_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-5_S16_L008_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T5-5/QF-1499-T5-5_S16_L008_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-5_S16_L008_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T5-70/QF-1499-T5-70_S18_L008_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-70_S18_L008_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T5-70/QF-1499-T5-70_S18_L008_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-70_S18_L008_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T5-81/QF-1499-T5-81_S13_L008_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-81_S13_L008_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T5-81/QF-1499-T5-81_S13_L008_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-81_S13_L008_R2_001.fastq.gz\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "866020bf-18dc-4987-96cb-3c48bed16775",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Prepare the reference fasta \n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/prepare_ref\n",
    "JOB_NAME=prepare_reference_fasta\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=3G\n",
    "#SBATCH --time=0-02:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "module load bwa/0.7.17\n",
    "module load java/jdk1.8.0_231\n",
    "module load picard/2.24.1\n",
    "module load htslib/1.15.1\n",
    "module load samtools/1.16.1\n",
    "\n",
    "\n",
    "bwa index -p '$SCRATCH_DIRECTORY'/ref_fasta/$(basename $REF_FASTA) '$SCRATCH_DIRECTORY'/ref_fasta/$(basename $REF_FASTA) -a bwtsw '$SCRATCH_DIRECTORY'/ref_fasta/$(basename $REF_FASTA)\n",
    "\n",
    "java -jar /gpfs/software/ada/picard/2.24.1/picard.jar CreateSequenceDictionary -R '$SCRATCH_DIRECTORY'/ref_fasta/$(basename $REF_FASTA) -O '$SCRATCH_DIRECTORY'/ref_fasta/$(basename $REF_FASTA '.fna.gz').dict\n",
    "\n",
    "zcat '$SCRATCH_DIRECTORY'/ref_fasta/$(basename $REF_FASTA) | bgzip -c > '$SCRATCH_DIRECTORY'/ref_fasta/$(basename $REF_FASTA)\n",
    "\n",
    "samtools faidx '$SCRATCH_DIRECTORY'/ref_fasta/$(basename $REF_FASTA)\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/prepare_ref.sh\n",
    "\n",
    "\n",
    "# sbatch $HOME_DIRECTORY/scripts/prepare_ref.sh\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8358c98b-7a7a-445d-8a4b-cb889ead5e63",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Generate the contig.intervals file for the reference fasta\n",
    "\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "\n",
    "\n",
    "cut -f1,2 $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA).fai > $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.temp\n",
    "\n",
    "less $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.temp | while read p; do echo \"1\"; done > $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.start\n",
    "\n",
    "paste $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.temp $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.start | awk '{ print $1,$3,$2 }' > $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.interval_list_1\n",
    "\n",
    "sed -r 's/\\s+/:/' $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.interval_list_1 > $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.interval_list_2\n",
    "\n",
    "sed -r 's/\\s+/-/' $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.interval_list_2 > $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.intervals\n",
    "\n",
    "\n",
    "rm $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.temp\n",
    "rm $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.start\n",
    "rm $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.interval_list_1\n",
    "rm $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.interval_list_2\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf2d98d3-0408-4878-afdd-eddf72a23f74",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Unzip the reference assembly for accuMUlate\n",
    "\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "\n",
    "zcat $SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz > $SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b141d163-e3ed-4582-baa9-c4a56e42545d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Trim adapters for each sample\n",
    "\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/trimming\n",
    "JOB_NAME=trimming\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=6G\n",
    "#SBATCH --time=0-02:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "input_file_1=$1\n",
    "input_file_2=$2\n",
    "\n",
    "module load fastp/0.23.2\n",
    "\n",
    "fastp -i '$SCRATCH_DIRECTORY'/fastq_files/$input_file_1 -I '$SCRATCH_DIRECTORY'/fastq_files/$input_file_2 -o '$SCRATCH_DIRECTORY'/trimmed_seq/$(basename $input_file_1 '.fastq.gz')_fastp.fastq.gz -O '$SCRATCH_DIRECTORY'/trimmed_seq/$(basename $input_file_2 '.fastq.gz')_fastp.fastq.gz --detect_adapter_for_pe -q 15 -l 15\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/trimming.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/fastq_list | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/trimming.sh \"$a\" \"$b\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/trimming_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/trimming_submit_jobs\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c10926ed-8f09-4c04-bc48-0dc417f97126",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Map reads for each sample\n",
    "\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/bwa_mem\n",
    "JOB_NAME=bwa_mem\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=18G\n",
    "#SBATCH --time=0-72:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "input_file_1=$1\n",
    "input_file_2=$2\n",
    "output_file=$3\n",
    "\n",
    "module load bwa/0.7.17\n",
    "\n",
    "bwa mem -t 6 '$REF_FASTA' '$SCRATCH_DIRECTORY'/trimmed_seq/$(basename $input_file_1 '.fastq.gz')_fastp.fastq.gz '$SCRATCH_DIRECTORY'/trimmed_seq/$(basename $input_file_2 '.fastq.gz')_fastp.fastq.gz > '$SCRATCH_DIRECTORY'/trimmed_seq/$(basename $output_file).sam\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/bwa_mem.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/fastq_list | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/bwa_mem.sh \"$a\" \"$b\" \"$c\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/bwa_mem_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/bwa_mem_submit_jobs\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdae9ebc-1efb-4378-9946-60a6376ff081",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Sort the sam files\n",
    "\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/samtools_sort\n",
    "JOB_NAME=samtools_sort_sam_to_bam\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=18G\n",
    "#SBATCH --time=0-72:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "input_file=$1\n",
    "\n",
    "module load samtools/1.16.1\n",
    "\n",
    "samtools sort -T '$SCRATCH_DIRECTORY'/trimmed_seq/$(basename $input_file) -o '$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file).bam '$SCRATCH_DIRECTORY'/trimmed_seq/$(basename $input_file).sam \n",
    "' > $HOME_DIRECTORY/scripts/samtools_sort_sam_to_bam.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/fastq_list | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/samtools_sort_sam_to_bam.sh \"$c\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/samtools_sort_sam_to_bam_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/samtools_sort_sam_to_bam_submit_jobs\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5ef1c54-1ae3-467f-9885-8a121b19fc2d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## View bam files\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/samtools_view\n",
    "JOB_NAME=samtools_view_bam\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=18G\n",
    "#SBATCH --time=0-48:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "input_file=$1\n",
    "\n",
    "module load samtools/1.16.1\n",
    "\n",
    "samtools view '$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file).bam -h -b -o '$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file)_viewed.bam\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/samtools_view_bam.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/fastq_list | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/samtools_view_bam.sh \"$c\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/samtools_view_bam_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/samtools_view_bam_submit_jobs\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "768a40cb-7e9a-47d8-939b-cc81fa83694f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## MarkDuplicates\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/MarkDuplicates\n",
    "JOB_NAME=MarkDuplicates\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=36G\n",
    "#SBATCH --time=0-02:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "input_file=$1\n",
    "\n",
    "module load java/jdk1.8.0_231\n",
    "module load picard/2.24.1\n",
    "\n",
    "java -jar /gpfs/software/ada/picard/2.24.1/picard.jar MarkDuplicates I='$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file)_viewed.bam O='$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file)_viewed_marked.bam METRICS_FILE='$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file)_metrics.txt ASSUME_SORTED=true VALIDATION_STRINGENCY=SILENT\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/MarkDuplicates.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/fastq_list | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/MarkDuplicates.sh \"$c\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/MarkDuplicates_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/MarkDuplicates_submit_jobs\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61437844-c7d8-4c2b-8c92-771aad93544e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## AddOrReplaceReadGroups\n",
    "\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/AddOrReplaceReadGroups\n",
    "JOB_NAME=AddOrReplaceReadGroups\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=6G\n",
    "#SBATCH --time=0-12:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "input_file=$1\n",
    "\n",
    "module load java/jdk1.8.0_231\n",
    "module load picard/2.24.1\n",
    "\n",
    "java -jar /gpfs/software/ada/picard/2.24.1/picard.jar AddOrReplaceReadGroups I='$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file)_viewed_marked.bam  O='$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file)_viewed_marked_addg.bam RGID=$(basename $input_file) RGLB=lib1 RGPL=illumina_TruSeq RGSM=$(basename $input_file) RGPU=unit1\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/AddOrReplaceReadGroups.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/fastq_list | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/AddOrReplaceReadGroups.sh \"$c\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/AddOrReplaceReadGroups_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/AddOrReplaceReadGroups_submit_jobs\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "808247ef-cc9e-447e-a4a5-73e78ffef0e3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Index the bam files\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/samtools_index\n",
    "JOB_NAME=samtools_index_bam\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=6G\n",
    "#SBATCH --time=0-01:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "input_file=$1\n",
    "\n",
    "module load samtools/1.16.1\n",
    "\n",
    "samtools index '$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file)_viewed_marked_addg.bam \n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/samtools_index.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/fastq_list | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/samtools_index.sh \"$c\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/samtools_index_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/samtools_index_submit_jobs\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2c08761-4024-4c46-a468-06f87d07d3ed",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## HaplotypeCaller\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/HaplotypeCaller\n",
    "JOB_NAME=HaplotypeCaller\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=30G\n",
    "#SBATCH --time=6-00:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "input_file=$1\n",
    "\n",
    "module load java/jdk1.8.0_231\n",
    "module load GATK/4.9.1\n",
    "\n",
    "gatk HaplotypeCaller -R '$REF_FASTA' -I '$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file)_viewed_marked_addg.bam -O '$HOME_DIRECTORY'/vcf_files/$(basename $input_file).g.vcf.gz -ploidy 2 -ERC BP_RESOLUTION --dont-use-soft-clipped-bases\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/HaplotypeCaller.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/fastq_list | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/HaplotypeCaller.sh \"$c\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/HaplotypeCaller_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/HaplotypeCaller_submit_jobs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13a7dd92-4329-4f9e-8739-b74f15caeb34",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## GenomicsDBImport\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "REF_FASTA_INTERVALS=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz_contigs.intervals\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/GenomicsDBImport\n",
    "JOB_NAME=GenomicsDBImport\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=80G\n",
    "#SBATCH --time=0-24:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "treatment=$1\n",
    "\n",
    "module load java/jdk1.8.0_231\n",
    "module load GATK/4.9.1\n",
    "\n",
    "gatk GenomicsDBImport --genomicsdb-workspace-path '$HOME_DIRECTORY'/vcf_handling/GenomicsDBImport_database/$(basename $treatment)_database --tmp-dir '$HOME_DIRECTORY'/vcf_handling/GenomicsDBImport_temp/$(basename $treatment)_temp --batch-size 10 --sample-name-map '$HOME_DIRECTORY'/$(basename $treatment)_sample_map --reader-threads 10 -L '$REF_FASTA_INTERVALS' --showHidden True --merge-input-intervals True\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/GenomicsDBImport.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/treatment_list | while read a; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/GenomicsDBImport.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/GenomicsDBImport_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/GenomicsDBImport_submit_jobs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d405b6f-513e-4891-9f18-8588e5c0ac09",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Joint genotyping for each treatment using GenotypeGVCFs; this produces a multi-sample vcf for each treatment that contains all samples in the treatment\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/GenotypeGVCFs\n",
    "JOB_NAME=GenotypeGVCFs\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=30G\n",
    "#SBATCH --time=0-24:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "treatment=$1\n",
    "\n",
    "module load java/jdk1.8.0_231\n",
    "module load GATK/4.9.1\n",
    "\n",
    "gatk GenotypeGVCFs -R '$REF_FASTA' -V gendb://'$HOME_DIRECTORY'/vcf_handling/GenomicsDBImport_database/$(basename $treatment)_database -G StandardAnnotation -O '$HOME_DIRECTORY'/vcf_handling/GenotypeGVCFs/$(basename $treatment)_GenotypeGVCFs.g.vcf.gz --include-non-variant-sites\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/GenotypeGVCFs.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/treatment_list | while read a; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/GenotypeGVCFs.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/GenotypeGVCFs_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/GenotypeGVCFs_submit_jobs\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6482920a-1f56-4660-a821-2adffd67959e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Use bcftools annotate to derive DP for each sample\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/bcftools_annotate\n",
    "JOB_NAME=bcftools_annotate\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=6G\n",
    "#SBATCH --time=0-24:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "treatment=$1\n",
    "\n",
    "module load bcftools/1.15.1\n",
    "\n",
    "bcftools annotate -x ^FORMAT/DP '$HOME_DIRECTORY'/vcf_handling/GenotypeGVCFs/$(basename $treatment)_GenotypeGVCFs.g.vcf.gz -o '$HOME_DIRECTORY'/vcf_handling/bcftools_annotate/$(basename $treatment)_bcftools_annotate_DP\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/bcftools_annotate.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/treatment_list | while read a; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/bcftools_annotate.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/bcftools_annotate_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/bcftools_annotate_submit_jobs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6dc706f-d234-452d-9282-ac5a032b3a4d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Derive mean DP for each sample for callability and filtering\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/DP_range\n",
    "JOB_NAME=DP_range\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=3G\n",
    "#SBATCH --time=0-12:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "treatment=$1\n",
    "sample_ID=$2\n",
    "field=$3\n",
    "\n",
    "sed '/^#/d' '$HOME_DIRECTORY'/vcf_handling/bcftools_annotate/$(basename $treatment)_bcftools_annotate_DP | cut -f$field > '$HOME_DIRECTORY'/vcf_handling/bcftools_annotate/$(basename $sample_ID)_bcftools_annotate_DP \n",
    "\n",
    "awk '\"'\"'{ total += $1 } END { print total/NR }'\"'\"' '$HOME_DIRECTORY'/vcf_handling/bcftools_annotate/$(basename $sample_ID)_bcftools_annotate_DP  > '$HOME_DIRECTORY'/vcf_handling/bcftools_annotate/$(basename $sample_ID)_bcftools_annotate_DP_mean\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/DP_range.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/treatment_sample_field_for_DP | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/DP_range.sh \"$a\" \"$b\" \"$c\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/DP_range_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/DP_range_submit_jobs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c120c825-099f-458d-9006-3881dcc2dc8b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Generate a table containing the DP ranges of all samples for callability and filtering\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "cat $HOME_DIRECTORY/vcf_handling/bcftools_annotate/*mean | paste $HOME_DIRECTORY/treatment_sample_field_for_DP - > $HOME_DIRECTORY/vcf_handling/bcftools_annotate/DP_range\n",
    "\n",
    "awk '{$5=$4*0.5; $6=$4*2; print $0}' $HOME_DIRECTORY/vcf_handling/bcftools_annotate/DP_range > $HOME_DIRECTORY/vcf_handling/bcftools_annotate/DP_range_temp && mv $HOME_DIRECTORY/vcf_handling/bcftools_annotate/DP_range_temp $HOME_DIRECTORY/DP_range\n",
    "\n",
    "awk '{$7=int($5); $8=int($6); print $0}' $HOME_DIRECTORY/DP_range > $HOME_DIRECTORY/DP_range_temp && mv $HOME_DIRECTORY/DP_range_temp $HOME_DIRECTORY/DP_range\n",
    "\n",
    "# awk '{$9=XXX}'\n",
    "# $HOME_DIRECTORY/DP_range2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acf0967a-3707-4748-8c8c-b62af6ed64f3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Use SelectVariants to select SNP\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/SelectVariants_SNP\n",
    "JOB_NAME=SelectVariants_SNP\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=12G\n",
    "#SBATCH --time=00-06:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "treatment=$1\n",
    "\n",
    "module load java/jdk1.8.0_231\n",
    "module load GATK/4.9.1\n",
    "\n",
    "gatk SelectVariants -R '$REF_FASTA' -V '$HOME_DIRECTORY'/vcf_handling/GenotypeGVCFs/$(basename $treatment)_GenotypeGVCFs.g.vcf.gz -O '$HOME_DIRECTORY'/vcf_handling/SelectVariants_SNP/$(basename $treatment)_SelectVariants_SNP.g.vcf.gz --select-type-to-include SNP --select-type-to-include NO_VARIATION --exclude-non-variants false\n",
    "' > $HOME_DIRECTORY/scripts/SelectVariants_SNP.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/treatment_list | while read a; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/SelectVariants_SNP.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/SelectVariants_SNP_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/SelectVariants_SNP_submit_jobs\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45459994-9e95-472d-8667-76148ed7f28b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Use bcftools split to split the multi-sample vcf to six single-sample vcf\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/bcftools_split\n",
    "JOB_NAME=bcftools_split\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=3G\n",
    "#SBATCH --time=0-04:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "treatment=$1\n",
    "treatment_sample=$2\n",
    "\n",
    "module load bcftools/1.15.1\n",
    "module load htslib/1.15.1\n",
    "\n",
    "bcftools +split '$HOME_DIRECTORY'/vcf_handling/SelectVariants_SNP/$(basename $treatment)_SelectVariants_SNP.g.vcf.gz -O z -o '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $treatment)/\n",
    "\n",
    "tabix -p vcf '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $treatment)/$(basename $treatment_sample).vcf.gz\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/bcftools_split.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/treatment_sample_field_for_DP | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/bcftools_split.sh \"$a\" \"$b\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/bcftools_split_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/bcftools_split_submit_jobs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23777e02-e2fd-49c6-a606-ec31dac64c6d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Use bcftools view to filter each single-sample vcf based on the sample's DP range\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/bcftools_DP\n",
    "JOB_NAME=bcftools_view_DP\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=3G\n",
    "#SBATCH --time=0-02:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "treatment=$1\n",
    "treatment_sample=$2\n",
    "min_DP=$3\n",
    "max_DP=$4\n",
    "\n",
    "module load bcftools/1.15.1\n",
    "module load htslib/1.15.1\n",
    "\n",
    "bcftools view -i '\"'\"'FMT/DP > '\"'\"'$min_DP'\"'\"'  & FMT/DP < '\"'\"'$max_DP'\"'\"' '\"'\"' '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $treatment)/$(basename $treatment_sample).vcf.gz -o '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $treatment)/$(basename $treatment_sample)_DP.vcf.gz\n",
    "\n",
    "tabix -p vcf '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $treatment)/$(basename $treatment_sample)_DP.vcf.gz\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/bcftools_view_DP.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/DP_range | while read a b c d e f g h; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/bcftools_view_DP.sh \"$a\" \"$b\" \"$g\" \"$h\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/bcftools_view_DP_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/bcftools_view_DP_submit_jobs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4082c9ef-cd76-4246-9969-40e1e7c6c4ae",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Use bcftools isec to derive a multi-sample vcf for each treatment, which contains SNPs that passed the sample-specific DP filters for all samples within a treatment\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/bcftools_DP\n",
    "JOB_NAME=bcftools_intersect\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=18G\n",
    "#SBATCH --time=0-8:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "treatment=$1\n",
    "treatment_sample1=$2\n",
    "treatment_sample2=$3\n",
    "treatment_sample3=$4\n",
    "treatment_sample4=$5\n",
    "treatment_sample5=$6\n",
    "treatment_sample6=$7\n",
    "\n",
    "module load bcftools/1.15.1\n",
    "module load htslib/1.15.1\n",
    "\n",
    "\n",
    "# 1. generate the intersection within treatment for each treatment; this produces a multi-sample vcf containing sites that passed the sample-specific DP filters for all samples within a treatment\n",
    "\n",
    "bcftools isec -n=6 -w1 '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $treatment)/$(basename $treatment_sample1)_DP.vcf.gz '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $treatment)/$(basename $treatment_sample2)_DP.vcf.gz '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $treatment)/$(basename $treatment_sample3)_DP.vcf.gz '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $treatment)/$(basename $treatment_sample4)_DP.vcf.gz '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $treatment)/$(basename $treatment_sample5)_DP.vcf.gz '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $treatment)/$(basename $treatment_sample6)_DP.vcf.gz -o '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $treatment)/$(basename $treatment)_bcftools_isec.vcf.gz -O z\n",
    "\n",
    "tabix -p vcf '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $treatment)/$(basename $treatment)_bcftools_isec.vcf.gz\n",
    "\n",
    "# 2. generate the intersection between 1 and the multi-sample vcf containing SNPs generated by SelectVariants\n",
    "\n",
    "bcftools isec -n=2 -w1 '$HOME_DIRECTORY'/vcf_handling/SelectVariants_SNP/$(basename $treatment)_SelectVariants_SNP.g.vcf.gz '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $treatment)/$(basename $treatment)_bcftools_isec.vcf.gz -o '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $treatment)/$(basename $treatment)_SNP_DPfiltered.vcf.gz -O z\n",
    "\n",
    "tabix -p vcf '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $treatment)/$(basename $treatment)_SNP_DPfiltered.vcf.gz\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/bcftools_intersect.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/treatment_sample_wide | while read a b c d e f g; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/bcftools_intersect.sh \"$a\" \"$b\" \"$c\" \"$d\" \"$e\" \"$f\" \"$g\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/bcftools_intersect_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/bcftools_intersect_submit_jobs\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbd123c4-7ab3-4b9f-a863-6889ff63bf03",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Use gatk SelectVariants, gatk VariantFiltration and bcftools view to filter for biallelic SNPs where all samples within a treatment are called and passed the hard-filters (five hard-filters with different parameter thresholds are designed); use VariantsToTable to extract INFO and FORMAT fields from the resultant vcf to a table\n",
    "\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/VariantFiltration\n",
    "JOB_NAME=VariantFiltration\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=12G\n",
    "#SBATCH --time=0-06:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "\n",
    "treatment=$1\n",
    "parameter_combination=$2\n",
    "QD_threshold=$3\n",
    "FS_threshold=$4\n",
    "SOR_threshold=$5\n",
    "MQ_threshold=$6\n",
    "MQRankSum_threshold=$7\n",
    "ReadPosRankSum_threshold=$8\n",
    "\n",
    "\n",
    "module load java/jdk1.8.0_231\n",
    "module load GATK/4.9.1\n",
    "\n",
    "\n",
    "# 1. Filters to derive biallelic SNP sites where all samples within the treatment are called\n",
    "\n",
    "gatk SelectVariants -R '$REF_FASTA' \\\n",
    "-V '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $treatment)/$(basename $treatment)_SNP_DPfiltered.vcf.gz \\\n",
    "--max-nocall-number 0 \\\n",
    "--restrict-alleles-to BIALLELIC \\\n",
    "-O '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $treatment)/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC.vcf\n",
    "\n",
    "# 2. Filters to derive sites that passed the hard-filters\n",
    "\n",
    "gatk VariantFiltration -R '$REF_FASTA' \\\n",
    "-V '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $treatment)/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC.vcf \\\n",
    "-O '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination).vcf.gz \\\n",
    "--filter \"QD < $QD_threshold\" \\\n",
    "--filter-name \"QD_threshold\" \\\n",
    "--filter \"FS > $FS_threshold\" \\\n",
    "--filter-name \"FS_threshold\" \\\n",
    "--filter \"SOR > $SOR_threshold\" \\\n",
    "--filter-name \"SOR_threshold\" \\\n",
    "--filter \"MQ < $MQ_threshold\" \\\n",
    "--filter-name \"MQ_threshold\" \\\n",
    "--filter \"MQRankSum < $MQRankSum_threshold\" \\\n",
    "--filter-name \"MQRankSum_threshold\" \\\n",
    "--filter \"ReadPosRankSum < $ReadPosRankSum_threshold\" \\\n",
    "--filter-name \"ReadPosRankSum_threshold\" \n",
    "\n",
    "module load bcftools/1.15.1\n",
    "module load htslib/1.15.1\n",
    "\n",
    "bcftools view -i '\"'\"'FILTER=\"PASS\"'\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination).vcf.gz -o '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS.vcf.gz\n",
    "\n",
    "tabix -p vcf '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS.vcf.gz\n",
    "\n",
    "\n",
    "# 3. VariantsToTable\n",
    "\n",
    "gatk VariantsToTable -R '$REF_FASTA' \\\n",
    "-V '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS.vcf.gz \\\n",
    "-O '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS.table \\\n",
    "-F CHROM -F POS -F REF -F ALT -F QD -F QUAL -F DP \\\n",
    "-F MQRankSum -F ReadPosRankSum -F MQ -F FS -F SOR -F AC -F AF -F AN -F HOM-REF -F HET -F HOM-VAR -F NO-CALL -F TRANSITION -F TYPE -F VAR \\\n",
    "-GF GT -GF AD -GF DP -GF GQ \n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/VariantFiltration.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/VariantFiltration_parameter_combinations | while read a b c d e f g h; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/VariantFiltration.sh \"$a\" \"$b\" \"$c\" \"$d\" \"$e\" \"$f\" \"$g\" \"$h\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/VariantFiltration_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/VariantFiltration_submit_jobs\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e15549cb-ac94-417b-8d9d-49dca54283d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Split columns: GT and AD\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/VariantFiltration\n",
    "JOB_NAME=split_columns\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=3G\n",
    "#SBATCH --time=0-01:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "\n",
    "treatment=$1\n",
    "parameter_combination=$2\n",
    "\n",
    "\n",
    "## split GT columns\n",
    "\n",
    "sed '\"'\"'s/|/\\t/g'\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS.table > '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_temp1.table \n",
    "\n",
    "sed '\"'\"'s/\\//\\t/g'\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_temp1.table > '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_temp2.table \n",
    "\n",
    "## split AD columns\n",
    "\n",
    "sed '\"'\"'s/,/\\t/g'\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_temp2.table > '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited.table \n",
    "\n",
    "rm '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_temp1.table\n",
    "rm '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_temp2.table\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/split_columns.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/VariantFiltration_parameter_combinations | while read a b c d e f g h; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/split_columns.sh \"$a\" \"$b\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/split_columns_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/split_columns_submit_jobs\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd208916-f4d0-4601-aaa6-407622ca82dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Derive singletons\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/VariantFiltration\n",
    "JOB_NAME=singletons\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=3G\n",
    "#SBATCH --time=0-01:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "\n",
    "treatment=$1\n",
    "parameter_combination=$2\n",
    "\n",
    "## HOM-REF == 5, HET == 1\n",
    "\n",
    "awk '\"'\"'($16 == 5) && ($17 == 1) && ($18 == 0) {print $0}'\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited.table > '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited_singletons_HOMREF_HET.table\n",
    "\n",
    "## HOM-VAR == 5, HET == 1\n",
    "\n",
    "awk '\"'\"'($16 == 0) && ($17 == 1) && ($18 == 5) {print $0}'\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited.table > '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited_singletons_HOMVAR_HET.table\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/singletons.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/VariantFiltration_parameter_combinations | while read a b c d e f g h; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/singletons.sh \"$a\" \"$b\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/singletons_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/singletons_submit_jobs\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5d43965-8dcb-4e64-8fee-6c3e5f04d143",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Filtering to derive singletons where 1. the allelic ratio of mutant allele is between 0.25 and 0.75, and 2. the depth for the mutant allele = 0 in all non-focal samples\n",
    "\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/de_novo_mutation\n",
    "JOB_NAME=singletons_filtering\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=3G\n",
    "#SBATCH --time=0-01:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "\n",
    "treatment=$1\n",
    "parameter_combination=$2\n",
    "HOMREF_or_HOMVAR=$3\n",
    "treatment_sample1=$4\n",
    "treatment_sample2=$5\n",
    "treatment_sample3=$6\n",
    "treatment_sample4=$7\n",
    "treatment_sample5=$8\n",
    "treatment_sample6=$9\n",
    "\n",
    "\n",
    "## for sample1, HOM-REF/HET\n",
    "\n",
    "awk '\"'\"' ($23 == $3) && ($24 == $4) && ($26/($26+$25) >= 0.25) && ($26/($26+$25) <= 0.75) && ($32 == 0) && ($38 == 0) && ($44 == 0) && ($50 == 0) && ($56 == 0) {print $0} '\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited_singletons_$(basename $HOMREF_or_HOMVAR).table > '$HOME_DIRECTORY'/de_novo_mutation/$(basename $treatment_sample1)_$(basename $parameter_combination)_$(basename $HOMREF_or_HOMVAR).table\n",
    "\n",
    "## for sample1, HOM-VAR/HET\n",
    "\n",
    "awk '\"'\"' ($23 == $3) && ($24 == $4) && ($26/($26+$25) >= 0.25) && ($26/($26+$25) <= 0.75) && ($31 == 0) && ($37 == 0) && ($43 == 0) && ($49 == 0) && ($55 == 0) {print $0} '\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited_singletons_$(basename $HOMREF_or_HOMVAR).table > '$HOME_DIRECTORY'/de_novo_mutation/$(basename $treatment_sample1)_$(basename $parameter_combination)_$(basename $HOMREF_or_HOMVAR).table\n",
    "\n",
    "\n",
    "## for sample2, HOM-REF/HET\n",
    "\n",
    "awk '\"'\"' ($29 == $3) && ($30 == $4) && ($32/($32+$31) >= 0.25) && ($32/($32+$31) <= 0.75) && ($26 == 0) && ($38 == 0) && ($44 == 0) && ($50 == 0) && ($56 == 0) {print $0} '\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited_singletons_$(basename $HOMREF_or_HOMVAR).table > '$HOME_DIRECTORY'/de_novo_mutation/$(basename $treatment_sample2)_$(basename $parameter_combination)_$(basename $HOMREF_or_HOMVAR).table\n",
    "\n",
    "## for sample2, HOM-VAR/HET\n",
    "\n",
    "awk '\"'\"' ($29 == $3) && ($30 == $4) && ($32/($32+$31) >= 0.25) && ($32/($32+$31) <= 0.75) && ($25 == 0) && ($37 == 0) && ($43 == 0) && ($49 == 0) && ($55 == 0) {print $0} '\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited_singletons_$(basename $HOMREF_or_HOMVAR).table > '$HOME_DIRECTORY'/de_novo_mutation/$(basename $treatment_sample2)_$(basename $parameter_combination)_$(basename $HOMREF_or_HOMVAR).table\n",
    "\n",
    "## for sample3, HOM-REF/HET\n",
    "\n",
    "awk '\"'\"' ($35 == $3) && ($36 == $4) && ($38/($38+$37) >= 0.25) && ($38/($38+$37) <= 0.75) && ($26 == 0) && ($32 == 0) && ($44 == 0) && ($50 == 0) && ($56 == 0) {print $0} '\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited_singletons_$(basename $HOMREF_or_HOMVAR).table > '$HOME_DIRECTORY'/de_novo_mutation/$(basename $treatment_sample3)_$(basename $parameter_combination)_$(basename $HOMREF_or_HOMVAR).table\n",
    "\n",
    "## for sample3, HOM-VAR/HET\n",
    "\n",
    "awk '\"'\"' ($35 == $3) && ($36 == $4) && ($38/($38+$37) >= 0.25) && ($38/($38+$37) <= 0.75) && ($25 == 0) && ($31 == 0) && ($43 == 0) && ($49 == 0) && ($55 == 0) {print $0} '\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited_singletons_$(basename $HOMREF_or_HOMVAR).table > '$HOME_DIRECTORY'/de_novo_mutation/$(basename $treatment_sample3)_$(basename $parameter_combination)_$(basename $HOMREF_or_HOMVAR).table\n",
    "\n",
    "\n",
    "## for sample4, HOM-REF/HET\n",
    "\n",
    "awk '\"'\"' ($41 == $3) && ($42 == $4) && ($44/($44+$43) >= 0.25) && ($44/($44+$43) <= 0.75) && ($26 == 0) && ($32 == 0) && ($38 == 0) && ($50 == 0) && ($56 == 0) {print $0} '\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited_singletons_$(basename $HOMREF_or_HOMVAR).table > '$HOME_DIRECTORY'/de_novo_mutation/$(basename $treatment_sample4)_$(basename $parameter_combination)_$(basename $HOMREF_or_HOMVAR).table\n",
    "\n",
    "## for sample4, HOM-VAR/HET\n",
    "\n",
    "awk '\"'\"' ($41 == $3) && ($42 == $4) && ($44/($44+$43) >= 0.25) && ($44/($44+$43) <= 0.75) && ($25 == 0) && ($31 == 0) && ($37 == 0) && ($49 == 0) && ($55 == 0) {print $0} '\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited_singletons_$(basename $HOMREF_or_HOMVAR).table > '$HOME_DIRECTORY'/de_novo_mutation/$(basename $treatment_sample4)_$(basename $parameter_combination)_$(basename $HOMREF_or_HOMVAR).table\n",
    "\n",
    "\n",
    "## for sample5, HOM-REF/HET\n",
    "\n",
    "awk '\"'\"' ($47 == $3) && ($48 == $4) && ($50/($49+$50) >= 0.25) && ($50/($49+$50) <= 0.75) && ($26 == 0) && ($32 == 0) && ($38 == 0) && ($44 == 0) && ($56 == 0) {print $0} '\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited_singletons_$(basename $HOMREF_or_HOMVAR).table > '$HOME_DIRECTORY'/de_novo_mutation/$(basename $treatment_sample5)_$(basename $parameter_combination)_$(basename $HOMREF_or_HOMVAR).table\n",
    "\n",
    "## for sample5, HOM-VAR/HET\n",
    "\n",
    "awk '\"'\"' ($47 == $3) && ($48 == $4) && ($50/($49+$50) >= 0.25) && ($50/($49+$50) <= 0.75) && ($25 == 0) && ($31 == 0) && ($37 == 0) && ($43 == 0) && ($55 == 0) {print $0} '\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited_singletons_$(basename $HOMREF_or_HOMVAR).table > '$HOME_DIRECTORY'/de_novo_mutation/$(basename $treatment_sample5)_$(basename $parameter_combination)_$(basename $HOMREF_or_HOMVAR).table\n",
    "\n",
    "\n",
    "## for sample6, HOM-REF/HET\n",
    "\n",
    "awk '\"'\"' ($53 == $3) && ($54 == $4) && ($56/($55+$56) >= 0.25) && ($56/($55+$56) <= 0.75) && ($26 == 0) && ($32 == 0) && ($38 == 0) && ($44 == 0) && ($50 == 0) {print $0} '\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited_singletons_$(basename $HOMREF_or_HOMVAR).table > '$HOME_DIRECTORY'/de_novo_mutation/$(basename $treatment_sample6)_$(basename $parameter_combination)_$(basename $HOMREF_or_HOMVAR).table\n",
    "\n",
    "## for sample6, HOM-VAR/HET\n",
    "\n",
    "awk '\"'\"' ($53 == $3) && ($54 == $4) && ($56/($55+$56) >= 0.25) && ($56/($55+$56) <= 0.75) && ($25 == 0) && ($31 == 0) && ($37 == 0) && ($43 == 0) && ($49 == 0) {print $0} '\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $treatment)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited_singletons_$(basename $HOMREF_or_HOMVAR).table > '$HOME_DIRECTORY'/de_novo_mutation/$(basename $treatment_sample6)_$(basename $parameter_combination)_$(basename $HOMREF_or_HOMVAR).table ' > $HOME_DIRECTORY/scripts/singletons_filtering.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/treatment_parameter_type_sample_wide | while read a b c d e f g h i; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/singletons_filtering.sh \"$a\" \"$b\" \"$c\" \"$d\" \"$e\" \"$f\" \"$g\" \"$h\" \"$i\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/singletons_filtering_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/singletons_filtering_submit_jobs\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5666c1bf-3bde-43ec-adc9-b58a8e9e4cd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## generate a table to store the results (number of lines)\n",
    "\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "\n",
    "touch $HOME_DIRECTORY/de_novo_mutation/results\n",
    "\n",
    "less $HOME_DIRECTORY/treatment_parameter_type_sample_wide | while read a b c d e f g h i; do printf \"\n",
    "wc -l \"$d\"_\"$b\"_\"$c\".table >> $HOME_DIRECTORY/de_novo_mutation/results\n",
    "wc -l \"$e\"_\"$b\"_\"$c\".table >> $HOME_DIRECTORY/de_novo_mutation/results\n",
    "wc -l \"$f\"_\"$b\"_\"$c\".table >> $HOME_DIRECTORY/de_novo_mutation/results\n",
    "wc -l \"$g\"_\"$b\"_\"$c\".table >> $HOME_DIRECTORY/de_novo_mutation/results\n",
    "wc -l \"$h\"_\"$b\"_\"$c\".table >> $HOME_DIRECTORY/de_novo_mutation/results\n",
    "wc -l \"$i\"_\"$b\"_\"$c\".table >> $HOME_DIRECTORY/de_novo_mutation/results\n",
    "\"; done > $HOME_DIRECTORY/scripts/HaplotypeCaller_results\n",
    "\n",
    "sh -e $HOME_DIRECTORY/scripts/HaplotypeCaller_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "092d542f-c3b2-42bb-9b90-b6718c71bd09",
   "metadata": {},
   "outputs": [],
   "source": [
    "## callability: split the GenotypeGVCFs file to single-sample vcf\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/callability\n",
    "JOB_NAME=callability_split\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=6G\n",
    "#SBATCH --time=0-24:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "treatment=$1\n",
    "\n",
    "module load bcftools/1.15.1\n",
    "\n",
    "bcftools +split '$HOME_DIRECTORY'/vcf_handling/GenotypeGVCFs/$(basename $treatment)_GenotypeGVCFs.g.vcf.gz -O z -o '$HOME_DIRECTORY'/vcf_handling/callability/$(basename $treatment)/\n",
    "\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/callability_split.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/treatment_list | while read a; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/callability_split.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/callability_split_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/callability_split_submit_jobs\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2300a01-5c10-4c50-967f-46357eb36962",
   "metadata": {},
   "outputs": [],
   "source": [
    "## callability: individually\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/callability\n",
    "JOB_NAME=callability_individual\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=3G\n",
    "#SBATCH --time=0-02:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "treatment=$1\n",
    "treatment_sample=$2\n",
    "field=$3\n",
    "\n",
    "module load bcftools/1.15.1\n",
    "\n",
    "bcftools view -i '\"'\"' FORMAT/GT['\"'\"'$field'\"'\"'] != \"mis\" '\"'\"' \\\n",
    "'$HOME_DIRECTORY'/vcf_handling/GenotypeGVCFs/$(basename $treatment)_GenotypeGVCFs.g.vcf.gz \\\n",
    "--output '$HOME_DIRECTORY'/vcf_handling/callability/$(basename $treatment_sample)_GenotypeGVCFs_callable.vcf.gz\n",
    "\n",
    "bcftools index -t '$HOME_DIRECTORY'/vcf_handling/callability/$(basename $treatment_sample)_GenotypeGVCFs_callable.vcf.gz\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/callability_individual.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/treatment_sample_field_for_callability | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/callability_individual.sh \"$a\" \"$b\" \"$c\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/callability_individual_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/callability_individual_submit_jobs\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab3491db-e1c0-417d-b570-c6826af72a4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## callability: individually\n",
    "## generate a table to store the results (number of lines)\n",
    "\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "\n",
    "touch $HOME_DIRECTORY/vcf_handling/callability/callability_individual_result\n",
    "\n",
    "less $HOME_DIRECTORY/treatment_sample_field_for_DP | while read a b c; do printf \"\n",
    "bcftools index --nrecords $HOME_DIRECTORY/vcf_handling/callability/\"$b\"_GenotypeGVCFs_callable.vcf.gz >> $HOME_DIRECTORY/vcf_handling/callability/callability_individual_result\n",
    "\"; done > $HOME_DIRECTORY/scripts/callability_individual_derive_results\n",
    "\n",
    "\n",
    "sed -i '1i module load bcftools/1.15.1' $HOME_DIRECTORY/scripts/callability_individual_derive_results\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/callability_individual_derive_results\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "607b3aeb-9ad3-4623-97d7-2bed41002081",
   "metadata": {},
   "outputs": [],
   "source": [
    "## callability: cohort\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/callability\n",
    "JOB_NAME=callability_cohort\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=3G\n",
    "#SBATCH --time=0-04:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "treatment=$1\n",
    "treatment_sample=$2\n",
    "\n",
    "module load bcftools/1.15.1\n",
    "\n",
    "bcftools view -i '\"'\"' FORMAT/GT[0] != \"mis\" && FORMAT/GT[1] != \"mis\" && FORMAT/GT[2] != \"mis\"  && FORMAT/GT[3] != \"mis\" && FORMAT/GT[4] != \"mis\" && FORMAT/GT[5] != \"mis\" '\"'\"' '$HOME_DIRECTORY'/vcf_handling/GenotypeGVCFs/$(basename $treatment)_GenotypeGVCFs.g.vcf.gz \\\n",
    "--output '$HOME_DIRECTORY'/vcf_handling/callability/$(basename $treatment)_GenotypeGVCFs_callable.g.vcf.gz\n",
    "\n",
    "bcftools index -t '$HOME_DIRECTORY'/vcf_handling/callability/$(basename $treatment)_GenotypeGVCFs_callable.g.vcf.gz\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/callability_cohort.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/treatment_list | while read a b; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/callability_cohort.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/callability_cohort_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/callability_cohort_submit_jobs\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d47f3fe-3794-4f42-ba09-8d0b8bbb5fe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "## callability: cohort\n",
    "## generate a table to store the results (number of lines)\n",
    "\n",
    "\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "\n",
    "touch $HOME_DIRECTORY/vcf_handling/callability/callability_individual_result\n",
    "\n",
    "less $HOME_DIRECTORY/treatment_list | while read a; do printf \"\n",
    "bcftools index --nrecords $HOME_DIRECTORY/vcf_handling/callability/\"$a\"_GenotypeGVCFs_callable.g.vcf.gz >> $HOME_DIRECTORY/vcf_handling/callability/callability_cohort_result\n",
    "\"; done > $HOME_DIRECTORY/scripts/callability_cohort_derive_results\n",
    "\n",
    "\n",
    "sed -i '1i module load bcftools/1.15.1' $HOME_DIRECTORY/scripts/callability_cohort_derive_results\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/callability_cohort_derive_results\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50b9acd7-2544-48e0-a95c-fd0e4beef011",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28f56aff-8f8e-4b2b-b0fd-71d78bcccc63",
   "metadata": {},
   "outputs": [],
   "source": [
    "################## accuMUlate ##################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a08258e0-4559-43a3-bc68-e78b56cb998a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## samtools merge\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/samtools_merge\n",
    "JOB_NAME=samtools_merge\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=60G\n",
    "#SBATCH --time=0-48:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "treatment=$1\n",
    "treatment_sample1=$2\n",
    "treatment_sample2=$3\n",
    "treatment_sample3=$4\n",
    "treatment_sample4=$5\n",
    "treatment_sample5=$6\n",
    "treatment_sample6=$7\n",
    "\n",
    "module load samtools/1.16.1\n",
    "\n",
    "samtools merge -o '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_merged.bam -r \\\n",
    "'$SCRATCH_DIRECTORY'/bam_files/$(basename $treatment_sample1)_viewed_marked_addg.bam \\\n",
    "'$SCRATCH_DIRECTORY'/bam_files/$(basename $treatment_sample2)_viewed_marked_addg.bam \\\n",
    "'$SCRATCH_DIRECTORY'/bam_files/$(basename $treatment_sample3)_viewed_marked_addg.bam \\\n",
    "'$SCRATCH_DIRECTORY'/bam_files/$(basename $treatment_sample4)_viewed_marked_addg.bam \\\n",
    "'$SCRATCH_DIRECTORY'/bam_files/$(basename $treatment_sample5)_viewed_marked_addg.bam \\\n",
    "'$SCRATCH_DIRECTORY'/bam_files/$(basename $treatment_sample6)_viewed_marked_addg.bam \n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/samtools_merge.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/treatment_sample_wide | while read a b c d e f g; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/samtools_merge.sh \"$a\" \"$b\" \"$c\" \"$d\" \"$e\" \"$f\" \"$g\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/samtools_merge_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/samtools_merge_submit_jobs\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f751b92e-1361-4f76-abf7-48b4ef270db0",
   "metadata": {},
   "outputs": [],
   "source": [
    "## picard MergeSamFiles \n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output\n",
    "JOB_NAME=picard_MergeSamFiles\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=60G\n",
    "#SBATCH --time=0-48:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "treatment=$1\n",
    "treatment_sample1=$2\n",
    "treatment_sample2=$3\n",
    "treatment_sample3=$4\n",
    "treatment_sample4=$5\n",
    "treatment_sample5=$6\n",
    "treatment_sample6=$7\n",
    "\n",
    "module load java/jdk1.8.0_231\n",
    "module load picard/2.24.1\n",
    "\n",
    "\n",
    "java -jar /gpfs/software/ada/picard/2.24.1/picard.jar MergeSamFiles \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $treatment_sample1)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $treatment_sample2)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $treatment_sample3)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $treatment_sample4)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $treatment_sample5)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $treatment_sample6)_viewed_marked_addg.bam \\\n",
    "O='$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_MergeSamFiles.bam\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/MergeSamFiles.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/treatment_sample_wide | while read a b c d e f g; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/MergeSamFiles.sh \"$a\" \"$b\" \"$c\" \"$d\" \"$e\" \"$f\" \"$g\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/MergeSamFiles_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/MergeSamFiles_submit_jobs\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da78620d-6654-403d-8218-3022dc20cc65",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Add a readgroup for the ancestor in the header of the nmerged bam file\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/accuMUlate\n",
    "JOB_NAME=addg\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=12G\n",
    "#SBATCH --time=0-12:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "treatment=$1\n",
    "\n",
    "\n",
    "module load samtools/1.16.1\n",
    "\n",
    "samtools view -H '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_MergeSamFiles.bam > '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_MergeSamFiles.head.sam\n",
    "\n",
    "sed '\"'\"'19 i @RG\\tID:Ancestral\\tLB:lib1 PL:illumina_TruSeq\\tSM:Ancestral\\tPU:unit1'\"'\"' '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_MergeSamFiles.head.sam > '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_MergeSamFiles_addg.head.sam\n",
    "\n",
    "samtools view '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_MergeSamFiles.bam | cat '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_MergeSamFiles_addg.head.sam - | samtools view -Sb - > '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_MergeSamFiles_addg.bam\n",
    "\n",
    "samtools index '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_MergeSamFiles_addg.bam\n",
    "\n",
    "\n",
    "module load python/3.10.0\n",
    "\n",
    "samtools view -H '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_MergeSamFiles_addg.bam | python '$HOME_DIRECTORY'/extract_samples.py Ancestral - >> '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_MergeSamFiles_params.ini\n",
    "\n",
    "\n",
    "cat '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_MergeSamFiles_params.ini '$HOME_DIRECTORY'/accuMUlate_parameters > '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_MergeSamFiles_params_temp.ini && mv '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_MergeSamFiles_params_temp.ini '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_MergeSamFiles_params.ini\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/addg.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/treatment_list | while read a; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/addg.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/addg_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/addg_submit_jobs\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88821155-6f2c-4c1c-befc-38aa89a29590",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Run accuMUlate\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "REF_FASTA_UNCOMPRESSED=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/accuMUlate\n",
    "JOB_NAME=accuMUlate\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=30G\n",
    "#SBATCH --time=5-00:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "treatment=$1\n",
    "\n",
    "module load accuMUlate/0.2.2\n",
    "\n",
    "/gpfs/software/ada/accuMUlate/0.2.2/accuMUlate \\\n",
    "-b '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_MergeSamFiles_addg.bam \\\n",
    "-r '$REF_FASTA_UNCOMPRESSED' \\\n",
    "-q 13 \\\n",
    "-m 13 \\\n",
    "-p 0.10000000000000001 \\\n",
    "--ploidy-ancestor 2 \\\n",
    "--ploidy-descendant 2 \\\n",
    "-c '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_MergeSamFiles_params.ini \\\n",
    "-o '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_putative_mutations.tsv\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/accuMUlate.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/treatment_list | while read a b c d e f g; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/accuMUlate.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/accuMUlate_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/accuMUlate_submit_jobs\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86e5a443-dd10-4bd8-b4ed-6ec17e0a8bf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Derive denominator\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/accuMUlate\n",
    "JOB_NAME=denominator\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=60G\n",
    "#SBATCH --time=0-72:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "treatment=$1\n",
    "\n",
    "module load accuMUlate/0.2.2\n",
    "\n",
    "/gpfs/software/ada/accuMUlate/0.2.2/denominate -c '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_denom_params.ini -r '$REF_FASTA_UNCOMPRESSED' -b '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_MergeSamFiles_addg.bam > '$HOME_DIRECTORY'/accuMUlate/$(basename $treatment)_denom.out\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/denominator.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/treatment_list | while read a; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/denominator.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/denominator_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/denominator_submit_jobs\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "toc-autonumbering": true,
  "toc-showcode": true,
  "toc-showmarkdowntxt": false,
  "toc-showtags": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
