{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a97d6827-e99b-402c-a4a9-8f3a32513903",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"0\"></a> <br>\n",
    " # Table of Contents  \n",
    "1. [Settings](#1)     \n",
    "    1. [Generate directories](#2)\n",
    "    1. [Define variables](#3)\n",
    "    1. [Generate symbolic links](#4)\n",
    "1. [Prepare the reference fasta](#5)\n",
    "    1. [Index the reference fasta](#6)\n",
    "    1. [Generate the contig.intervals file](#7)\n",
    "    1. [Decompress the reference fasta for the probabilistic approach](#8)\n",
    "1. [Pre-processing](#9)\n",
    "    1. [Trimming](#10)\n",
    "        1. [Trim adapters](#11)\n",
    "    1. [Mapping](#12)\n",
    "        1. [Map adapter-trimmed reads against the reference assembly](#13)\n",
    "        1. [Sort the sam files](#14)\n",
    "        1. [View the bam files](#15)\n",
    "        1. [Mark duplicated reads in the bam files](#16)\n",
    "        1. [Add read groups to the bam files](#17)\n",
    "        1. [Index the bam files](#18)\n",
    "1. [Variant calling by the consensus approach](#19)\n",
    "    1. [Variant calling](#20)\n",
    "        1. [Call variants](#21)\n",
    "        1. [Import variants to a GenomicsDB](#22)\n",
    "        1. [Joint genotyping](#23)\n",
    "    1. [Derive the coverage limits for each sample](#24)\n",
    "        1. [Derive coverage for each site for each sample](#25)\n",
    "        1. [Calculate mean coverage across sites for each sample](#26)\n",
    "        1. [Calculate the upper and lower coverage limits for each sample](#27)\n",
    "    1. [Variant filtering](#28)\n",
    "        1. [Derive SNP sites](#29)\n",
    "        1. [Coverage filtering](#30)\n",
    "            1. [Split the multi-sample vcf to single-sample vcf](#31)\n",
    "            1. [Coverage filtering for each sample](#32)\n",
    "            1. [Intersect to derive a multi-sample vcf for each regime](#33)\n",
    "        1. [Further filtering and hard-filtering](#34)\n",
    "        1. [Derive heterozygous singletons](#35)\n",
    "            1. [Split GT and AD columns](#36)\n",
    "            1. [Derive singletons](#37)\n",
    "            1. [Filter on the allelic ratio of the mutant allele](#38)\n",
    "        1. [Derive the number of candidate germline mutations for each sample](#39)\n",
    "        1. [Derive bed files for manual inspection](#40)\n",
    "    1. [Callability](#41)\n",
    "        1. [Derive genotyped sites for each regime](#42)\n",
    "        1. [Derive the number of callable sites for each regime](#43)\n",
    "1. [Variant calling by the probabilistic approach](#44)\n",
    "    1. [Merge bam files for each regime](#45)\n",
    "    1. [Add a readgroup for the ancestor to the header of the merged bam](#46)\n",
    "    1. [Call variants](#47)\n",
    "    1. [Variant filtering](#48)\n",
    "    1. [Derive the number of candidate germline mutations for each sample](#49)\n",
    "    1. [Derive denominator](#50)\n",
    "    1. [Derive bed files for manual inspection](#51)\n",
    "1. [Manual inspection](#52)\n",
    "    1. [Merge and annotate the bed files from both approaches for each sample](#53)\n",
    "    1. [Derive single-sample vcf](#54)\n",
    "    1. [Index single-sample vcf](#55)\n",
    "    1. [Files needed for manual inspection](#56)\n",
    "    1. [Find overlapping variants](#57)\n",
    "1. [Aggregate the results](#58)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d0a07b8-a475-4105-b21d-f3983f9de9e9",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"1\"></a> \n",
    "# Settings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d11d1844-819d-4497-953f-5795d9af4d4c",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"2\"></a> \n",
    "## Generate directories\n",
    "This section generates the directories that will be used for the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17bffff9-5c89-4285-be6f-23a1a8d2189c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "mkdir $HOME_DIRECTORY/scripts\n",
    "\n",
    "mkdir $SCRATCH_DIRECTORY/job_output\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/prepare_ref/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/trimming/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/bwa_mem/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/samtools_sort/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/samtools_view/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/MarkDuplicates/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/AddOrReplaceReadGroups/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/samtools_index/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/HaplotypeCaller/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/GenomicsDBImport/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/GenotypeGVCFs/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/bcftools_annotate/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/DP_range/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/SelectVariants_SNP/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/bcftools_split/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/bcftools_DP/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/VariantFiltration/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/callability\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/samtools_merge/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/accuMUlate/\n",
    "mkdir $SCRATCH_DIRECTORY/job_output/de_novo_mutation/\n",
    "\n",
    "\n",
    "mkdir $SCRATCH_DIRECTORY/ref_fasta/\n",
    "mkdir $SCRATCH_DIRECTORY/fastq_files/\n",
    "mkdir $SCRATCH_DIRECTORY/trimmed_seq/\n",
    "mkdir $SCRATCH_DIRECTORY/bam_files/\n",
    "mkdir $HOME_DIRECTORY/vcf_files/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/callability/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/callability/T1/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/callability/T2/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/callability/T5/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/GenomicsDBImport_database/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/GenomicsDBImport_temp/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/GenomicsDBImport_temp/T1_temp\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/GenomicsDBImport_temp/T2_temp\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/GenomicsDBImport_temp/T5_temp\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/GenotypeGVCFs/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/bcftools_annotate/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/SelectVariants_SNP/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/bcftools_DP/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/bcftools_DP/T1/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/bcftools_DP/T2/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/bcftools_DP/T5/\n",
    "mkdir $HOME_DIRECTORY/vcf_handling/VariantFiltration/\n",
    "mkdir $HOME_DIRECTORY/de_novo_mutation/\n",
    "mkdir $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\n",
    "mkdir $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach/\n",
    "mkdir $HOME_DIRECTORY/de_novo_mutation/bed/\n",
    "mkdir $HOME_DIRECTORY/de_novo_mutation/bed/manual_inspection/\n",
    "mkdir $HOME_DIRECTORY/accuMUlate/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5119a8c3-0760-4e33-a3d4-97f1b20c49cb",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"3\"></a> \n",
    "## Define variables \n",
    "This section defines the variables that will be used for the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d97dff75-c17a-46f9-8163-49d10376f6f0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "echo \"QF-1499-T1-10_S14_L008_R1_001.fastq.gz QF-1499-T1-10_S14_L008_R2_001.fastq.gz T1-10\n",
    "QF-1499-T1-19_S1_L001_R1_001.fastq.gz QF-1499-T1-19_S1_L001_R2_001.fastq.gz T1-19\n",
    "QF-1499-T1-37_S2_L001_R1_001.fastq.gz QF-1499-T1-37_S2_L001_R2_001.fastq.gz T1-37\n",
    "QF-1499-T1-46_S15_L008_R1_001.fastq.gz QF-1499-T1-46_S15_L008_R2_001.fastq.gz T1-46\n",
    "QF-1499-T1-57_S3_L001_R1_001.fastq.gz QF-1499-T1-57_S3_L001_R2_001.fastq.gz T1-57\n",
    "QF-1499-T1-6_S17_L008_R1_001.fastq.gz QF-1499-T1-6_S17_L008_R2_001.fastq.gz T1-6\n",
    "QF-1499-T2-1_S4_L001_R1_001.fastq.gz QF-1499-T2-1_S4_L001_R2_001.fastq.gz T2-1\n",
    "QF-1499-T2-22_S6_L001_R1_001.fastq.gz QF-1499-T2-22_S6_L001_R2_001.fastq.gz T2-22\n",
    "QF-1499-T2-34_S12_L008_R1_001.fastq.gz QF-1499-T2-34_S12_L008_R2_001.fastq.gz T2-34\n",
    "QF-1499-T2-4_S5_L001_R1_001.fastq.gz QF-1499-T2-4_S5_L001_R2_001.fastq.gz T2-4\n",
    "QF-1499-T2-42_S11_L008_R1_001.fastq.gz QF-1499-T2-42_S11_L008_R2_001.fastq.gz T2-42\n",
    "QF-1499-T2-52_S7_L001_R1_001.fastq.gz QF-1499-T2-52_S7_L001_R2_001.fastq.gz T2-52\n",
    "QF-1499-T5-45_S10_L008_R1_001.fastq.gz QF-1499-T5-45_S10_L008_R2_001.fastq.gz T5-45\n",
    "QF-1499-T5-5_S16_L008_R1_001.fastq.gz QF-1499-T5-5_S16_L008_R2_001.fastq.gz T5-5\n",
    "QF-1499-T5-70_S18_L008_R1_001.fastq.gz QF-1499-T5-70_S18_L008_R2_001.fastq.gz T5-70\n",
    "QF-1499-T5-81_S13_L008_R1_001.fastq.gz QF-1499-T5-81_S13_L008_R2_001.fastq.gz T5-81\n",
    "QF-1499-T5-86_S8_L001_R1_001.fastq.gz QF-1499-T5-86_S8_L001_R2_001.fastq.gz T5-86\n",
    "QF-1499-T5-87_S9_L001_R1_001.fastq.gz QF-1499-T5-87_S9_L001_R2_001.fastq.gz T5-87\"> $HOME_DIRECTORY/fastq_list\n",
    "\n",
    "\n",
    "echo \"T1\n",
    "T2\n",
    "T5\" > $HOME_DIRECTORY/regime_list\n",
    "\n",
    "\n",
    "echo \"T1-10\n",
    "T1-19\n",
    "T1-37\n",
    "T1-46\n",
    "T1-57\n",
    "T1-6\" > $HOME_DIRECTORY/T1_sample_list\n",
    "\n",
    "echo \"$HOME_DIRECTORY/vcf_files/T1-10.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T1-19.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T1-37.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T1-46.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T1-57.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T1-6.g.vcf.gz\" > $HOME_DIRECTORY/T1_vcf_list\n",
    "\n",
    "paste -d\"\\t\" $HOME_DIRECTORY/T1_sample_list $HOME_DIRECTORY/T1_vcf_list > $HOME_DIRECTORY/T1_sample_map\n",
    "\n",
    "echo \"T2-1\n",
    "T2-22\n",
    "T2-34\n",
    "T2-4\n",
    "T2-42\n",
    "T2-52\" > $HOME_DIRECTORY/T2_sample_list\n",
    "\n",
    "echo \"$HOME_DIRECTORY/vcf_files/T2-1.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T2-22.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T2-34.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T2-4.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T2-42.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T2-52.g.vcf.gz\" > $HOME_DIRECTORY/T2_vcf_list\n",
    "\n",
    "paste -d\"\\t\" $HOME_DIRECTORY/T2_sample_list $HOME_DIRECTORY/T2_vcf_list > $HOME_DIRECTORY/T2_sample_map\n",
    "\n",
    "echo \"T5-45\n",
    "T5-5\n",
    "T5-70\n",
    "T5-81\n",
    "T5-86\n",
    "T5-87\" > $HOME_DIRECTORY/T5_sample_list\n",
    "\n",
    "echo \"$HOME_DIRECTORY/vcf_files/T5-45.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T5-5.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T5-70.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T5-81.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T5-86.g.vcf.gz\n",
    "$HOME_DIRECTORY/vcf_files/T5-87.g.vcf.gz\" > $HOME_DIRECTORY/T5_vcf_list\n",
    "\n",
    "paste -d\"\\t\" $HOME_DIRECTORY/T5_sample_list $HOME_DIRECTORY/T5_vcf_list > $HOME_DIRECTORY/T5_sample_map\n",
    "\n",
    "\n",
    "printf \"T1 T1-10\n",
    "T1 T1-19\n",
    "T1 T1-37\n",
    "T1 T1-46\n",
    "T1 T1-57\n",
    "T1 T1-6\n",
    "T2 T2-1\n",
    "T2 T2-22\n",
    "T2 T2-34\n",
    "T2 T2-4\n",
    "T2 T2-42\n",
    "T2 T2-52\n",
    "T5 T5-45\n",
    "T5 T5-5\n",
    "T5 T5-70\n",
    "T5 T5-81\n",
    "T5 T5-86\n",
    "T5 T5-87\" > $HOME_DIRECTORY/regime_sample\n",
    "\n",
    "printf \"T1 T1-10 10\n",
    "T1 T1-19 11\n",
    "T1 T1-37 12\n",
    "T1 T1-46 13\n",
    "T1 T1-57 14\n",
    "T1 T1-6 15\n",
    "T2 T2-1 10\n",
    "T2 T2-22 11\n",
    "T2 T2-34 12\n",
    "T2 T2-4 13\n",
    "T2 T2-42 14\n",
    "T2 T2-52 15\n",
    "T5 T5-45 10\n",
    "T5 T5-5 11\n",
    "T5 T5-70 12\n",
    "T5 T5-81 13\n",
    "T5 T5-86 14\n",
    "T5 T5-87 15\" > $HOME_DIRECTORY/regime_sample_field_for_DP\n",
    "\n",
    "printf \"T1 T1-10 0\n",
    "T1 T1-19 1\n",
    "T1 T1-37 2\n",
    "T1 T1-46 3\n",
    "T1 T1-57 4\n",
    "T1 T1-6 5\n",
    "T2 T2-1 0\n",
    "T2 T2-22 1\n",
    "T2 T2-34 2\n",
    "T2 T2-4 3\n",
    "T2 T2-42 4\n",
    "T2 T2-52 5\n",
    "T5 T5-45 0\n",
    "T5 T5-5 1\n",
    "T5 T5-70 2\n",
    "T5 T5-81 3\n",
    "T5 T5-86 4\n",
    "T5 T5-87 5\" > $HOME_DIRECTORY/regime_sample_field_for_callability\n",
    "\n",
    "\n",
    "printf \"T1 T1-10 T1-19 T1-37 T1-46 T1-57 T1-6\n",
    "T2 T2-1 T2-22 T2-34 T2-4 T2-42 T2-52\n",
    "T5 T5-45 T5-5 T5-70 T5-81 T5-86 T5-87\" > $HOME_DIRECTORY/regime_sample_wide\n",
    "\n",
    "echo \"T1 1 2.00 60.00 3.00 40.00 -12.50 -8.00\n",
    "T1 2 2.00 20.00 3.00 40.00 -12.50 -8.00\n",
    "T1 3 2.00 60.00 2.00 40.00 -12.50 -8.00\n",
    "T1 4 2.00 60.00 2.00 50.00 -12.50 -8.00\n",
    "T1 5 2.00 60.00 3.00 40.00 -6.00 -8.00\n",
    "T1 6 2.00 60.00 3.00 40.00 -12.50 -4.00\n",
    "T2 1 2.00 60.00 3.00 40.00 -12.50 -8.00\n",
    "T2 2 2.00 20.00 3.00 40.00 -12.50 -8.00\n",
    "T2 3 2.00 60.00 2.00 40.00 -12.50 -8.00\n",
    "T2 4 2.00 60.00 2.00 50.00 -12.50 -8.00\n",
    "T2 5 2.00 60.00 3.00 40.00 -6.00 -8.00\n",
    "T2 6 2.00 60.00 3.00 40.00 -12.50 -4.00\n",
    "T5 1 2.00 60.00 3.00 40.00 -12.50 -8.00\n",
    "T5 2 2.00 20.00 3.00 40.00 -12.50 -8.00\n",
    "T5 3 2.00 60.00 2.00 40.00 -12.50 -8.00\n",
    "T5 4 2.00 60.00 2.00 50.00 -12.50 -8.00\n",
    "T5 5 2.00 60.00 3.00 40.00 -6.00 -8.00\n",
    "T5 6 2.00 60.00 3.00 40.00 -12.50 -4.00\" > $HOME_DIRECTORY/VariantFiltration_parameter_combinations\n",
    "\n",
    "\n",
    "echo \"T1 1 HOMREF_HET T1-10 T1-19 T1-37 T1-46 T1-57 T1-6\n",
    "T1 1 HOMVAR_HET T1-10 T1-19 T1-37 T1-46 T1-57 T1-6\n",
    "T1 2 HOMREF_HET T1-10 T1-19 T1-37 T1-46 T1-57 T1-6\n",
    "T1 2 HOMVAR_HET T1-10 T1-19 T1-37 T1-46 T1-57 T1-6\n",
    "T1 3 HOMREF_HET T1-10 T1-19 T1-37 T1-46 T1-57 T1-6\n",
    "T1 3 HOMVAR_HET T1-10 T1-19 T1-37 T1-46 T1-57 T1-6\n",
    "T1 4 HOMREF_HET T1-10 T1-19 T1-37 T1-46 T1-57 T1-6\n",
    "T1 4 HOMVAR_HET T1-10 T1-19 T1-37 T1-46 T1-57 T1-6\n",
    "T1 5 HOMREF_HET T1-10 T1-19 T1-37 T1-46 T1-57 T1-6\n",
    "T1 5 HOMVAR_HET T1-10 T1-19 T1-37 T1-46 T1-57 T1-6\n",
    "T1 6 HOMREF_HET T1-10 T1-19 T1-37 T1-46 T1-57 T1-6\n",
    "T1 6 HOMVAR_HET T1-10 T1-19 T1-37 T1-46 T1-57 T1-6\n",
    "T2 1 HOMREF_HET T2-1 T2-22 T2-34 T2-4 T2-42 T2-52\n",
    "T2 1 HOMVAR_HET T2-1 T2-22 T2-34 T2-4 T2-42 T2-52\n",
    "T2 2 HOMREF_HET T2-1 T2-22 T2-34 T2-4 T2-42 T2-52\n",
    "T2 2 HOMVAR_HET T2-1 T2-22 T2-34 T2-4 T2-42 T2-52\n",
    "T2 3 HOMREF_HET T2-1 T2-22 T2-34 T2-4 T2-42 T2-52\n",
    "T2 3 HOMVAR_HET T2-1 T2-22 T2-34 T2-4 T2-42 T2-52\n",
    "T2 4 HOMREF_HET T2-1 T2-22 T2-34 T2-4 T2-42 T2-52\n",
    "T2 4 HOMVAR_HET T2-1 T2-22 T2-34 T2-4 T2-42 T2-52\n",
    "T2 5 HOMREF_HET T2-1 T2-22 T2-34 T2-4 T2-42 T2-52\n",
    "T2 5 HOMVAR_HET T2-1 T2-22 T2-34 T2-4 T2-42 T2-52\n",
    "T2 6 HOMREF_HET T2-1 T2-22 T2-34 T2-4 T2-42 T2-52\n",
    "T2 6 HOMVAR_HET T2-1 T2-22 T2-34 T2-4 T2-42 T2-52\n",
    "T5 1 HOMREF_HET T5-45 T5-5 T5-70 T5-81 T5-86 T5-87\n",
    "T5 1 HOMVAR_HET T5-45 T5-5 T5-70 T5-81 T5-86 T5-87\n",
    "T5 2 HOMREF_HET T5-45 T5-5 T5-70 T5-81 T5-86 T5-87\n",
    "T5 2 HOMVAR_HET T5-45 T5-5 T5-70 T5-81 T5-86 T5-87\n",
    "T5 3 HOMREF_HET T5-45 T5-5 T5-70 T5-81 T5-86 T5-87\n",
    "T5 3 HOMVAR_HET T5-45 T5-5 T5-70 T5-81 T5-86 T5-87\n",
    "T5 4 HOMREF_HET T5-45 T5-5 T5-70 T5-81 T5-86 T5-87\n",
    "T5 4 HOMVAR_HET T5-45 T5-5 T5-70 T5-81 T5-86 T5-87\n",
    "T5 5 HOMREF_HET T5-45 T5-5 T5-70 T5-81 T5-86 T5-87\n",
    "T5 5 HOMVAR_HET T5-45 T5-5 T5-70 T5-81 T5-86 T5-87\n",
    "T5 6 HOMREF_HET T5-45 T5-5 T5-70 T5-81 T5-86 T5-87\n",
    "T5 6 HOMVAR_HET T5-45 T5-5 T5-70 T5-81 T5-86 T5-87\" > $HOME_DIRECTORY/regime_parameter_type_sample_wide\n",
    "\n",
    "\n",
    "\n",
    "echo \"T1\t1\tT1-10\n",
    "T1\t2\tT1-10\n",
    "T1\t3\tT1-10\n",
    "T1\t4\tT1-10\n",
    "T1\t5\tT1-10\n",
    "T1\t6\tT1-10\n",
    "T2\t1\tT2-1\n",
    "T2\t2\tT2-1\n",
    "T2\t3\tT2-1\n",
    "T2\t4\tT2-1\n",
    "T2\t5\tT2-1\n",
    "T2\t6\tT2-1\n",
    "T5\t1\tT5-45\n",
    "T5\t2\tT5-45\n",
    "T5\t3\tT5-45\n",
    "T5\t4\tT5-45\n",
    "T5\t5\tT5-45\n",
    "T5\t6\tT5-45\" > $HOME_DIRECTORY/regime_parameter_type_sample1\n",
    "\n",
    "echo \"T1\t1\tT1-19\n",
    "T1\t2\tT1-19\n",
    "T1\t3\tT1-19\n",
    "T1\t4\tT1-19\n",
    "T1\t5\tT1-19\n",
    "T1\t6\tT1-19\n",
    "T2\t1\tT2-22\n",
    "T2\t2\tT2-22\n",
    "T2\t3\tT2-22\n",
    "T2\t4\tT2-22\n",
    "T2\t5\tT2-22\n",
    "T2\t6\tT2-22\n",
    "T5\t1\tT5-5\n",
    "T5\t2\tT5-5\n",
    "T5\t3\tT5-5\n",
    "T5\t4\tT5-5\n",
    "T5\t5\tT5-5\n",
    "T5\t6\tT5-5\" > $HOME_DIRECTORY/regime_parameter_type_sample2\n",
    "\n",
    "echo \"T1\t1\tT1-37\n",
    "T1\t2\tT1-37\n",
    "T1\t3\tT1-37\n",
    "T1\t4\tT1-37\n",
    "T1\t5\tT1-37\n",
    "T1\t6\tT1-37\n",
    "T2\t1\tT2-34\n",
    "T2\t2\tT2-34\n",
    "T2\t3\tT2-34\n",
    "T2\t4\tT2-34\n",
    "T2\t5\tT2-34\n",
    "T2\t6\tT2-34\n",
    "T5\t1\tT5-70\n",
    "T5\t2\tT5-70\n",
    "T5\t3\tT5-70\n",
    "T5\t4\tT5-70\n",
    "T5\t5\tT5-70\n",
    "T5\t6\tT5-70\" > $HOME_DIRECTORY/regime_parameter_type_sample3\n",
    "\n",
    "echo \"T1\t1\tT1-46\n",
    "T1\t2\tT1-46\n",
    "T1\t3\tT1-46\n",
    "T1\t4\tT1-46\n",
    "T1\t5\tT1-46\n",
    "T1\t6\tT1-46\n",
    "T2\t1\tT2-4\n",
    "T2\t2\tT2-4\n",
    "T2\t3\tT2-4\n",
    "T2\t4\tT2-4\n",
    "T2\t5\tT2-4\n",
    "T2\t6\tT2-4\n",
    "T5\t1\tT5-81\n",
    "T5\t2\tT5-81\n",
    "T5\t3\tT5-81\n",
    "T5\t4\tT5-81\n",
    "T5\t5\tT5-81\n",
    "T5\t6\tT5-81\" > $HOME_DIRECTORY/regime_parameter_type_sample4\n",
    "\n",
    "echo \"T1\t1\tT1-57\n",
    "T1\t2\tT1-57\n",
    "T1\t3\tT1-57\n",
    "T1\t4\tT1-57\n",
    "T1\t5\tT1-57\n",
    "T1\t6\tT1-57\n",
    "T2\t1\tT2-42\n",
    "T2\t2\tT2-42\n",
    "T2\t3\tT2-42\n",
    "T2\t4\tT2-42\n",
    "T2\t5\tT2-42\n",
    "T2\t6\tT2-42\n",
    "T5\t1\tT5-86\n",
    "T5\t2\tT5-86\n",
    "T5\t3\tT5-86\n",
    "T5\t4\tT5-86\n",
    "T5\t5\tT5-86\n",
    "T5\t6\tT5-86\" > $HOME_DIRECTORY/regime_parameter_type_sample5\n",
    "\n",
    "echo \"T1\t1\tT1-6\n",
    "T1\t2\tT1-6\n",
    "T1\t3\tT1-6\n",
    "T1\t4\tT1-6\n",
    "T1\t5\tT1-6\n",
    "T1\t6\tT1-6\n",
    "T2\t1\tT2-52\n",
    "T2\t2\tT2-52\n",
    "T2\t3\tT2-52\n",
    "T2\t4\tT2-52\n",
    "T2\t5\tT2-52\n",
    "T2\t6\tT2-52\n",
    "T5\t1\tT5-87\n",
    "T5\t2\tT5-87\n",
    "T5\t3\tT5-87\n",
    "T5\t4\tT5-87\n",
    "T5\t5\tT5-87\n",
    "T5\t6\tT5-87\" > $HOME_DIRECTORY/regime_parameter_type_sample6\n",
    "\n",
    "\n",
    "echo \"theta=0.0001\n",
    "mu=0.00000001\n",
    "nfreqs=0.25     0.25    0.25    0.25\n",
    "seq-error=0.01\n",
    "ploidy-ancestor=2\n",
    "ploidy-descendant=2\n",
    "phi-diploid=0.001\n",
    "phi-haploid=0.001\n",
    "\" > $HOME_DIRECTORY/accuMUlate_parameters\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05d3a014-25b7-4fd4-9a30-7115ff87f880",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"4\"></a> \n",
    "## Generate symbolic links \n",
    "This section generates symbolic links for each sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e4d7b80-5334-47c0-a615-d688f0038e63",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T1-19/QF-1499-T1-19_S1_L001_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-19_S1_L001_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T1-19/QF-1499-T1-19_S1_L001_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-19_S1_L001_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T1-37/QF-1499-T1-37_S2_L001_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-37_S2_L001_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T1-37/QF-1499-T1-37_S2_L001_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-37_S2_L001_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T1-57/QF-1499-T1-57_S3_L001_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-57_S3_L001_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T1-57/QF-1499-T1-57_S3_L001_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-57_S3_L001_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T2-1/QF-1499-T2-1_S4_L001_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-1_S4_L001_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T2-1/QF-1499-T2-1_S4_L001_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-1_S4_L001_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T2-22/QF-1499-T2-22_S6_L001_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-22_S6_L001_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T2-22/QF-1499-T2-22_S6_L001_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-22_S6_L001_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T2-4/QF-1499-T2-4_S5_L001_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-4_S5_L001_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T2-4/QF-1499-T2-4_S5_L001_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-4_S5_L001_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T2-52/QF-1499-T2-52_S7_L001_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-52_S7_L001_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T2-52/QF-1499-T2-52_S7_L001_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-52_S7_L001_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T5-86/QF-1499-T5-86_S8_L001_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-86_S8_L001_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T5-86/QF-1499-T5-86_S8_L001_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-86_S8_L001_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T5-87/QF-1499-T5-87_S9_L001_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-87_S9_L001_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00346/Sample_QF-1499-T5-87/QF-1499-T5-87_S9_L001_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-87_S9_L001_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T1-10/QF-1499-T1-10_S14_L008_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-10_S14_L008_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T1-10/QF-1499-T1-10_S14_L008_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-10_S14_L008_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T1-46/QF-1499-T1-46_S15_L008_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-46_S15_L008_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T1-46/QF-1499-T1-46_S15_L008_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-46_S15_L008_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T1-6/QF-1499-T1-6_S17_L008_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-6_S17_L008_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T1-6/QF-1499-T1-6_S17_L008_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T1-6_S17_L008_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T2-34/QF-1499-T2-34_S12_L008_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-34_S12_L008_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T2-34/QF-1499-T2-34_S12_L008_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-34_S12_L008_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T2-42/QF-1499-T2-42_S11_L008_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-42_S11_L008_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T2-42/QF-1499-T2-42_S11_L008_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T2-42_S11_L008_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T5-45/QF-1499-T5-45_S10_L008_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-45_S10_L008_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T5-45/QF-1499-T5-45_S10_L008_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-45_S10_L008_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T5-5/QF-1499-T5-5_S16_L008_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-5_S16_L008_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T5-5/QF-1499-T5-5_S16_L008_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-5_S16_L008_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T5-70/QF-1499-T5-70_S18_L008_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-70_S18_L008_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T5-70/QF-1499-T5-70_S18_L008_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-70_S18_L008_R2_001.fastq.gz\n",
    "\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T5-81/QF-1499-T5-81_S13_L008_R1_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-81_S13_L008_R1_001.fastq.gz\n",
    "ln -s $FILE_DIRECTORY/nematode_data_QF1499_delivery00393/Sample_QF-1499-T5-81/QF-1499-T5-81_S13_L008_R2_001.fastq.gz $SCRATCH_DIRECTORY/fastq_files/QF-1499-T5-81_S13_L008_R2_001.fastq.gz\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5acbe8e3-fa5f-41fb-8408-e35393ba6fb5",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"5\"></a> \n",
    "# Prepare the reference fasta "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "deef2e41-d715-49e9-83cf-f4e9df9d2376",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"6\"></a> \n",
    "## Index the reference fasta "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "866020bf-18dc-4987-96cb-3c48bed16775",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/prepare_ref\n",
    "JOB_NAME=prepare_reference_fasta\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=3G\n",
    "#SBATCH --time=0-02:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "module load bwa/0.7.17\n",
    "module load java/jdk1.8.0_231\n",
    "module load picard/2.24.1\n",
    "module load htslib/1.15.1\n",
    "module load samtools/1.16.1\n",
    "\n",
    "\n",
    "bwa index -p '$SCRATCH_DIRECTORY'/ref_fasta/$(basename $REF_FASTA) '$SCRATCH_DIRECTORY'/ref_fasta/$(basename $REF_FASTA) -a bwtsw '$SCRATCH_DIRECTORY'/ref_fasta/$(basename $REF_FASTA)\n",
    "\n",
    "java -jar /gpfs/software/ada/picard/2.24.1/picard.jar CreateSequenceDictionary -R '$SCRATCH_DIRECTORY'/ref_fasta/$(basename $REF_FASTA) -O '$SCRATCH_DIRECTORY'/ref_fasta/$(basename $REF_FASTA '.fna.gz').dict\n",
    "\n",
    "zcat '$SCRATCH_DIRECTORY'/ref_fasta/$(basename $REF_FASTA) | bgzip -c > '$SCRATCH_DIRECTORY'/ref_fasta/$(basename $REF_FASTA)\n",
    "\n",
    "samtools faidx '$SCRATCH_DIRECTORY'/ref_fasta/$(basename $REF_FASTA)\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/prepare_ref.sh\n",
    "\n",
    "\n",
    "# sbatch $HOME_DIRECTORY/scripts/prepare_ref.sh\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab907a9d-1175-40f2-a359-0bcec67a54b1",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"7\"></a> \n",
    "## Generate the contig.intervals file\n",
    "This script generates the contig.intervals file that will be used in the [GenomicsDBImport](#22) step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8358c98b-7a7a-445d-8a4b-cb889ead5e63",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "\n",
    "\n",
    "cut -f1,2 $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA).fai > $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.temp\n",
    "\n",
    "less $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.temp | while read p; do echo \"1\"; done > $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.start\n",
    "\n",
    "paste $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.temp $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.start | awk '{ print $1,$3,$2 }' > $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.interval_list_1\n",
    "\n",
    "sed -r 's/\\s+/:/' $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.interval_list_1 > $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.interval_list_2\n",
    "\n",
    "sed -r 's/\\s+/-/' $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.interval_list_2 > $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.intervals\n",
    "\n",
    "\n",
    "rm $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.temp\n",
    "rm $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.start\n",
    "rm $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.interval_list_1\n",
    "rm $SCRATCH_DIRECTORY/ref_fasta/$(basename $REF_FASTA)_contigs.interval_list_2\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52765022-8868-4ce1-9256-5923ff32b6b6",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"8\"></a> \n",
    "## Decompress the reference fasta for the probabilistic approach\n",
    "This script produces a decompressed fastq for the reference assembly which will be used for the [probabilistic approach](#XXX)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf2d98d3-0408-4878-afdd-eddf72a23f74",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "FILE_DIRECTORY=/gpfs/data/justin_rna/Hwei-yen\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "\n",
    "zcat $SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz > $SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d8112c2-40f5-4941-9d4e-1f8d5707144e",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"9\"></a> \n",
    "# Pre-processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9b77b06-7a9b-4bd9-ba9a-76bdf278eb09",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"10\"></a> \n",
    "## Trimming"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03222a68-bf61-47a4-ba94-2ca45e07521b",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"11\"></a> \n",
    "### Trim adapters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b141d163-e3ed-4582-baa9-c4a56e42545d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/trimming\n",
    "JOB_NAME=trimming\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=6G\n",
    "#SBATCH --time=0-02:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "input_file_1=$1\n",
    "input_file_2=$2\n",
    "\n",
    "module load fastp/0.23.2\n",
    "\n",
    "fastp -i '$SCRATCH_DIRECTORY'/fastq_files/$input_file_1 -I '$SCRATCH_DIRECTORY'/fastq_files/$input_file_2 -o '$SCRATCH_DIRECTORY'/trimmed_seq/$(basename $input_file_1 '.fastq.gz')_fastp.fastq.gz -O '$SCRATCH_DIRECTORY'/trimmed_seq/$(basename $input_file_2 '.fastq.gz')_fastp.fastq.gz --detect_adapter_for_pe -q 15 -l 15\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/trimming.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/fastq_list | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/trimming.sh \"$a\" \"$b\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/trimming_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/trimming_submit_jobs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0d3c8cd-6f9a-4cd7-9ab1-84ea1dcd35bd",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"12\"></a> \n",
    "## Mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45de8065-974c-44a9-9cf2-31e4e2d03a2f",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"13\"></a> \n",
    "### Map adapter-trimmed reads against the reference assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c10926ed-8f09-4c04-bc48-0dc417f97126",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/bwa_mem\n",
    "JOB_NAME=bwa_mem\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=18G\n",
    "#SBATCH --time=0-72:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "input_file_1=$1\n",
    "input_file_2=$2\n",
    "output_file=$3\n",
    "\n",
    "module load bwa/0.7.17\n",
    "\n",
    "bwa mem -t 6 '$REF_FASTA' '$SCRATCH_DIRECTORY'/trimmed_seq/$(basename $input_file_1 '.fastq.gz')_fastp.fastq.gz '$SCRATCH_DIRECTORY'/trimmed_seq/$(basename $input_file_2 '.fastq.gz')_fastp.fastq.gz > '$SCRATCH_DIRECTORY'/trimmed_seq/$(basename $output_file).sam\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/bwa_mem.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/fastq_list | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/bwa_mem.sh \"$a\" \"$b\" \"$c\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/bwa_mem_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/bwa_mem_submit_jobs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30022242-6e63-4e78-8df9-1c5cf184940b",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"14\"></a> \n",
    "### Sort the sam files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdae9ebc-1efb-4378-9946-60a6376ff081",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/samtools_sort\n",
    "JOB_NAME=samtools_sort_sam_to_bam\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=18G\n",
    "#SBATCH --time=0-72:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "input_file=$1\n",
    "\n",
    "module load samtools/1.16.1\n",
    "\n",
    "samtools sort -T '$SCRATCH_DIRECTORY'/trimmed_seq/$(basename $input_file) -o '$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file).bam '$SCRATCH_DIRECTORY'/trimmed_seq/$(basename $input_file).sam \n",
    "' > $HOME_DIRECTORY/scripts/samtools_sort_sam_to_bam.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/fastq_list | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/samtools_sort_sam_to_bam.sh \"$c\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/samtools_sort_sam_to_bam_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/samtools_sort_sam_to_bam_submit_jobs\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c95baef-71d2-4565-bc05-41c108a85039",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"15\"></a> \n",
    "### View the bam files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5ef1c54-1ae3-467f-9885-8a121b19fc2d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/samtools_view\n",
    "JOB_NAME=samtools_view_bam\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=18G\n",
    "#SBATCH --time=0-48:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "input_file=$1\n",
    "\n",
    "module load samtools/1.16.1\n",
    "\n",
    "samtools view '$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file).bam -h -b -o '$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file)_viewed.bam\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/samtools_view_bam.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/fastq_list | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/samtools_view_bam.sh \"$c\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/samtools_view_bam_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/samtools_view_bam_submit_jobs\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "534c44e0-bec2-4912-8566-0721c8a87313",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"16\"></a> \n",
    "### Mark duplicated reads in the bam files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "768a40cb-7e9a-47d8-939b-cc81fa83694f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/MarkDuplicates\n",
    "JOB_NAME=MarkDuplicates\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=36G\n",
    "#SBATCH --time=0-02:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "input_file=$1\n",
    "\n",
    "module load java/jdk1.8.0_231\n",
    "module load picard/2.24.1\n",
    "\n",
    "java -jar /gpfs/software/ada/picard/2.24.1/picard.jar MarkDuplicates I='$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file)_viewed.bam O='$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file)_viewed_marked.bam METRICS_FILE='$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file)_metrics.txt ASSUME_SORTED=true VALIDATION_STRINGENCY=SILENT\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/MarkDuplicates.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/fastq_list | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/MarkDuplicates.sh \"$c\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/MarkDuplicates_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/MarkDuplicates_submit_jobs\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1763954c-5369-4e18-bb24-8533f8bccb33",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"17\"></a> \n",
    "### Add read groups to the bam files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61437844-c7d8-4c2b-8c92-771aad93544e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/AddOrReplaceReadGroups\n",
    "JOB_NAME=AddOrReplaceReadGroups\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=6G\n",
    "#SBATCH --time=0-12:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "input_file=$1\n",
    "\n",
    "module load java/jdk1.8.0_231\n",
    "module load picard/2.24.1\n",
    "\n",
    "java -jar /gpfs/software/ada/picard/2.24.1/picard.jar AddOrReplaceReadGroups I='$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file)_viewed_marked.bam  O='$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file)_viewed_marked_addg.bam RGID=$(basename $input_file) RGLB=lib1 RGPL=illumina_TruSeq RGSM=$(basename $input_file) RGPU=unit1\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/AddOrReplaceReadGroups.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/fastq_list | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/AddOrReplaceReadGroups.sh \"$c\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/AddOrReplaceReadGroups_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/AddOrReplaceReadGroups_submit_jobs\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8a660ff-0dee-45ec-8e36-e094052e2f61",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"18\"></a> \n",
    "### Index the bam files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "808247ef-cc9e-447e-a4a5-73e78ffef0e3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/samtools_index\n",
    "JOB_NAME=samtools_index_bam\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=6G\n",
    "#SBATCH --time=0-01:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "input_file=$1\n",
    "\n",
    "module load samtools/1.16.1\n",
    "\n",
    "samtools index '$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file)_viewed_marked_addg.bam \n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/samtools_index.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/fastq_list | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/samtools_index.sh \"$c\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/samtools_index_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/samtools_index_submit_jobs\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c05fa75b-ad80-41f1-a937-9defc4a9f6c9",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"19\"></a> \n",
    "# Variant calling by the consensus approach"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "422c9738-56f1-44d0-acbb-0dbb38d3da56",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"20\"></a> \n",
    "## Variant calling"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "435f1fa6-eb74-41e1-ad04-46226e2c9cb3",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"21\"></a> \n",
    "### Call variants\n",
    "This script uses the GATK HaplotypeCaller tool to generate a single-sample vcf for each sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2c08761-4024-4c46-a468-06f87d07d3ed",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/HaplotypeCaller\n",
    "JOB_NAME=HaplotypeCaller\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=30G\n",
    "#SBATCH --time=6-00:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "input_file=$1\n",
    "\n",
    "module load java/jdk1.8.0_231\n",
    "module load GATK/4.9.1\n",
    "\n",
    "gatk HaplotypeCaller -R '$REF_FASTA' -I '$SCRATCH_DIRECTORY'/bam_files/$(basename $input_file)_viewed_marked_addg.bam -O '$HOME_DIRECTORY'/vcf_files/$(basename $input_file).g.vcf.gz -ploidy 2 -ERC BP_RESOLUTION --dont-use-soft-clipped-bases\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/HaplotypeCaller.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/fastq_list | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/HaplotypeCaller.sh \"$c\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/HaplotypeCaller_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/HaplotypeCaller_submit_jobs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06ed822b-5a3f-4718-b951-b8f50149e075",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"22\"></a> \n",
    "### Import variants to a GenomicsDB\n",
    "This script uses the GATK GenomicsDBImport tool to import single-sample vcfs into a GenomicsDB before joint genotyping. Specifically, single-sample vcfs of the same regime are imported into a GenomicsDB (six single-sample vcfs per regime); total three GenomicsDBs, corresponding to the three regimes (T1, T2, T5), are generated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13a7dd92-4329-4f9e-8739-b74f15caeb34",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "REF_FASTA_INTERVALS=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz_contigs.intervals\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/GenomicsDBImport\n",
    "JOB_NAME=GenomicsDBImport\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=80G\n",
    "#SBATCH --time=0-24:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "\n",
    "module load java/jdk1.8.0_231\n",
    "module load GATK/4.9.1\n",
    "\n",
    "gatk GenomicsDBImport --genomicsdb-workspace-path '$HOME_DIRECTORY'/vcf_handling/GenomicsDBImport_database/$(basename $regime)_database --tmp-dir '$HOME_DIRECTORY'/vcf_handling/GenomicsDBImport_temp/$(basename $regime)_temp --batch-size 10 --sample-name-map '$HOME_DIRECTORY'/$(basename $regime)_sample_map --reader-threads 10 -L '$REF_FASTA_INTERVALS' --showHidden True --merge-input-intervals True\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/GenomicsDBImport.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_list | while read a; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/GenomicsDBImport.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/GenomicsDBImport_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/GenomicsDBImport_submit_jobs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6054230c-1863-402e-8609-722c581ccee5",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"23\"></a> \n",
    "### Joint genotyping\n",
    "This script uses the GATK GenotypeGVCFs tool to performe joint genotype and produces a multi-sample vcf for each GenomicsDB (i.e. for each regime)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d405b6f-513e-4891-9f18-8588e5c0ac09",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/GenotypeGVCFs\n",
    "JOB_NAME=GenotypeGVCFs\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=30G\n",
    "#SBATCH --time=0-24:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "\n",
    "module load java/jdk1.8.0_231\n",
    "module load GATK/4.9.1\n",
    "\n",
    "gatk GenotypeGVCFs -R '$REF_FASTA' -V gendb://'$HOME_DIRECTORY'/vcf_handling/GenomicsDBImport_database/$(basename $regime)_database -G StandardAnnotation -O '$HOME_DIRECTORY'/vcf_handling/GenotypeGVCFs/$(basename $regime)_GenotypeGVCFs.g.vcf.gz --include-non-variant-sites\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/GenotypeGVCFs.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_list | while read a; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/GenotypeGVCFs.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/GenotypeGVCFs_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/GenotypeGVCFs_submit_jobs\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52240abc-8550-4ba3-97fa-eeee9c506e18",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"24\"></a> \n",
    "## Derive the coverage limits for each sample\n",
    "This section calculates the mean coverage, as well as the uppler and lower coverage limits for each sample to facilitate [coverage filtering](#30)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8273102e-a260-46b8-a2d2-f28ec8d5970e",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"25\"></a> \n",
    "### Derive coverage for each site for each sample\n",
    "This script uses the bcftools annotate tool to derive the coverage (FORMAT/DP field) for each site for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6482920a-1f56-4660-a821-2adffd67959e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/bcftools_annotate\n",
    "JOB_NAME=bcftools_annotate\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=6G\n",
    "#SBATCH --time=0-24:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "\n",
    "module load bcftools/1.15.1\n",
    "\n",
    "bcftools annotate -x ^FORMAT/DP '$HOME_DIRECTORY'/vcf_handling/GenotypeGVCFs/$(basename $regime)_GenotypeGVCFs.g.vcf.gz -o '$HOME_DIRECTORY'/vcf_handling/bcftools_annotate/$(basename $regime)_bcftools_annotate_DP\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/bcftools_annotate.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_list | while read a; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/bcftools_annotate.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/bcftools_annotate_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/bcftools_annotate_submit_jobs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3bb0bf3-7f5e-465c-a16f-56d0e2fcca3b",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"26\"></a> \n",
    "### Calculate mean coverage across sites for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6dc706f-d234-452d-9282-ac5a032b3a4d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Derive mean DP for each sample\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/DP_range\n",
    "JOB_NAME=DP_range\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=3G\n",
    "#SBATCH --time=0-12:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "sample_ID=$2\n",
    "field=$3\n",
    "\n",
    "sed '/^#/d' '$HOME_DIRECTORY'/vcf_handling/bcftools_annotate/$(basename $regime)_bcftools_annotate_DP | cut -f$field > '$HOME_DIRECTORY'/vcf_handling/bcftools_annotate/$(basename $sample_ID)_bcftools_annotate_DP \n",
    "\n",
    "awk '\"'\"'{ total += $1 } END { print total/NR }'\"'\"' '$HOME_DIRECTORY'/vcf_handling/bcftools_annotate/$(basename $sample_ID)_bcftools_annotate_DP  > '$HOME_DIRECTORY'/vcf_handling/bcftools_annotate/$(basename $sample_ID)_bcftools_annotate_DP_mean\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/DP_range.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_sample_field_for_DP | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/DP_range.sh \"$a\" \"$b\" \"$c\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/DP_range_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/DP_range_submit_jobs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66c1af2b-9297-4f38-aaac-9dd335192463",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"27\"></a> \n",
    "### Calculate the upper and lower coverage limits for each sample\n",
    "This script calculates the upper (2 * mean coverage) and the lower (0.5 * mean coverage) limits for each sample, convert the numeric values to integer, and store the results in a table.\n",
    "\n",
    "**Interperating the table:** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c120c825-099f-458d-9006-3881dcc2dc8b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "cat $HOME_DIRECTORY/vcf_handling/bcftools_annotate/*mean | paste $HOME_DIRECTORY/regime_sample_field_for_DP - > $HOME_DIRECTORY/vcf_handling/bcftools_annotate/DP_range\n",
    "\n",
    "awk '{$5=$4*0.5; $6=$4*2; print $0}' $HOME_DIRECTORY/vcf_handling/bcftools_annotate/DP_range > $HOME_DIRECTORY/vcf_handling/bcftools_annotate/DP_range_temp && mv $HOME_DIRECTORY/vcf_handling/bcftools_annotate/DP_range_temp $HOME_DIRECTORY/DP_range\n",
    "\n",
    "awk '{$7=int($5); $8=int($6); print $0}' $HOME_DIRECTORY/DP_range > $HOME_DIRECTORY/DP_range_temp && mv $HOME_DIRECTORY/DP_range_temp $HOME_DIRECTORY/DP_range\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0fb680c-bcb4-4e30-ad44-f3b5cbb9ae21",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"28\"></a> \n",
    "## Variant filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff2e6b52-e8ba-4402-9922-c4632f8baba3",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"29\"></a> \n",
    "### Derive SNP sites\n",
    "This script uses the GATK SelectVariants tool to select sites that are SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acf0967a-3707-4748-8c8c-b62af6ed64f3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/SelectVariants_SNP\n",
    "JOB_NAME=SelectVariants_SNP\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=12G\n",
    "#SBATCH --time=00-06:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "\n",
    "module load java/jdk1.8.0_231\n",
    "module load GATK/4.9.1\n",
    "\n",
    "gatk SelectVariants -R '$REF_FASTA' -V '$HOME_DIRECTORY'/vcf_handling/GenotypeGVCFs/$(basename $regime)_GenotypeGVCFs.g.vcf.gz -O '$HOME_DIRECTORY'/vcf_handling/SelectVariants_SNP/$(basename $regime)_SelectVariants_SNP.g.vcf.gz --select-type-to-include SNP --select-type-to-include NO_VARIATION --exclude-non-variants false\n",
    "' > $HOME_DIRECTORY/scripts/SelectVariants_SNP.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_list | while read a; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/SelectVariants_SNP.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/SelectVariants_SNP_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/SelectVariants_SNP_submit_jobs\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc6d0503-ff9b-4539-b8fe-96b9e6535dc2",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"30\"></a> \n",
    "### Coverage filtering\n",
    "This script filters to derive SNPs with coverages within the upper or lower coverage limits for each sample."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8debcae9-71b6-4265-a77b-e20d60464bac",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"31\"></a> \n",
    "#### Split the multi-sample vcf to single-sample vcf\n",
    "This script uses the bcftools split tool to split the multi-sample vcf containing SNPs for each regime (i.e. the output from [SelectVariants](#29)) to six single-sample vcf; total 18 single-sample vcf files (six per regime) are generated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45459994-9e95-472d-8667-76148ed7f28b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/bcftools_split\n",
    "JOB_NAME=bcftools_split\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=3G\n",
    "#SBATCH --time=0-04:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "sample_ID=$2\n",
    "\n",
    "module load bcftools/1.15.1\n",
    "module load htslib/1.15.1\n",
    "\n",
    "bcftools +split '$HOME_DIRECTORY'/vcf_handling/SelectVariants_SNP/$(basename $regime)_SelectVariants_SNP.g.vcf.gz -O z -o '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $regime)/\n",
    "\n",
    "tabix -p vcf '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $regime)/$(basename $sample_ID).vcf.gz\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/bcftools_split.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_sample_field_for_DP | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/bcftools_split.sh \"$a\" \"$b\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/bcftools_split_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/bcftools_split_submit_jobs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "763ce40b-93a6-4d32-9496-8c43ff824af8",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"32\"></a> \n",
    "#### Coverage filtering for each sample\n",
    "This script uses the bcftools view tool to derive sites that pass the sample's coverage filters for each sample (i.e. sites with coverages within 0.5* and 2* of the mean coverage of the sample are kept; see [Calculate the upper and lower coverage limits for each sample](#27)). The output of this script is a single-sample vcf for each sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23777e02-e2fd-49c6-a606-ec31dac64c6d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/bcftools_DP\n",
    "JOB_NAME=bcftools_view_DP\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=3G\n",
    "#SBATCH --time=0-02:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "sample_ID=$2\n",
    "min_DP=$3\n",
    "max_DP=$4\n",
    "\n",
    "module load bcftools/1.15.1\n",
    "module load htslib/1.15.1\n",
    "\n",
    "bcftools view -i '\"'\"'FMT/DP > '\"'\"'$min_DP'\"'\"'  & FMT/DP < '\"'\"'$max_DP'\"'\"' '\"'\"' '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $regime)/$(basename $sample_ID).vcf.gz -o '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $regime)/$(basename $sample_ID)_DP.vcf.gz\n",
    "\n",
    "tabix -p vcf '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $regime)/$(basename $sample_ID)_DP.vcf.gz\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/bcftools_view_DP.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/DP_range | while read a b c d e f g h; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/bcftools_view_DP.sh \"$a\" \"$b\" \"$g\" \"$h\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/bcftools_view_DP_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/bcftools_view_DP_submit_jobs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e9d190f-d302-4278-9272-5b7114ba6c92",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"33\"></a> \n",
    "#### Intersect to derive a multi-sample vcf for each regime\n",
    "This script uses the bcftools isec tool to generate a multi-sample vcf for each regime containg SNPs that pass the sample-specific coverage filters for all six samples within the regime; total three multi-sample vcf files, corresponding to the three regimes (T1, T2, T5), are generated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4082c9ef-cd76-4246-9969-40e1e7c6c4ae",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/bcftools_DP\n",
    "JOB_NAME=bcftools_intersect\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=18G\n",
    "#SBATCH --time=0-8:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "sample1=$2\n",
    "sample2=$3\n",
    "sample3=$4\n",
    "sample4=$5\n",
    "sample5=$6\n",
    "sample6=$7\n",
    "\n",
    "module load bcftools/1.15.1\n",
    "module load htslib/1.15.1\n",
    "\n",
    "\n",
    "# 1. generate the intersection within regime for each regime; this produces a multi-sample vcf containing sites that passed the sample-specific DP filters for all samples within a regime\n",
    "\n",
    "bcftools isec -n=6 -w1 '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $regime)/$(basename $sample1)_DP.vcf.gz '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $regime)/$(basename $sample2)_DP.vcf.gz '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $regime)/$(basename $sample3)_DP.vcf.gz '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $regime)/$(basename $sample4)_DP.vcf.gz '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $regime)/$(basename $sample5)_DP.vcf.gz '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $regime)/$(basename $sample6)_DP.vcf.gz -o '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $regime)/$(basename $regime)_bcftools_isec.vcf.gz -O z\n",
    "\n",
    "tabix -p vcf '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $regime)/$(basename $regime)_bcftools_isec.vcf.gz\n",
    "\n",
    "# 2. generate the intersection between 1 and the multi-sample vcf containing SNPs generated by SelectVariants\n",
    "\n",
    "bcftools isec -n=2 -w1 '$HOME_DIRECTORY'/vcf_handling/SelectVariants_SNP/$(basename $regime)_SelectVariants_SNP.g.vcf.gz '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $regime)/$(basename $regime)_bcftools_isec.vcf.gz -o '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $regime)/$(basename $regime)_SNP_DPfiltered.vcf.gz -O z\n",
    "\n",
    "tabix -p vcf '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $regime)/$(basename $regime)_SNP_DPfiltered.vcf.gz\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/bcftools_intersect.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_sample_wide | while read a b c d e f g; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/bcftools_intersect.sh \"$a\" \"$b\" \"$c\" \"$d\" \"$e\" \"$f\" \"$g\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/bcftools_intersect_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/bcftools_intersect_submit_jobs\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04dec084-18c3-41c2-9e63-72bb66f66059",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"34\"></a> \n",
    "### Further filtering and hard-filtering\n",
    "This script uses the multi-sample vcf for each regime generated from the [coverage filtering step](#33) as input and further filters to derive biallelic sites where all samples of the same regime are genotyped (Step 1; see below) and for sites that pass the five hard-filter combinations (Step 2; see below), and outputs a tab-delimited table for each hard-filter combination (Step 3). Total 15 tables, five for each regime, are generated.\n",
    "\n",
    "Specifically, the three steps are:\n",
    "**Step 1:** Use GATK SelectVariants tool to derive biallelic sites where all samples of the same regime are genotyped. This step uses the multi-sample vcf for each regime from the [coverage filtering step](#33) and outputs a multi-sample vcf for each regime for Step 2.\n",
    "**Step 2:** Use GATK VariantFiltration and bcftools view tools to derive sites that pass the hard-filters. Total five hard-filter combinations are designed (defined [here](#3)). This step uses the multi-sample vcf generated in Step 1 as input and outputs five resultant multi-sample vcf files for each input vcf, with each of the resultant vcf corresponding to a hard-filter combination.\n",
    "**Step 3:** Use GATK VariantsToTable tool to extract INFO and FORMAT fields from the input vcf and outputs a tab-delimited table. This step uses the multi-sample vcf generated in Step 2 as input and outputs a tab-delimited table for each input vcf. Total five tables, each corresponding to a hard-filter combination, are generated for each regime. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbd123c4-7ab3-4b9f-a863-6889ff63bf03",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/VariantFiltration\n",
    "JOB_NAME=VariantFiltration\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=12G\n",
    "#SBATCH --time=0-06:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "\n",
    "regime=$1\n",
    "parameter_combination=$2\n",
    "QD_threshold=$3\n",
    "FS_threshold=$4\n",
    "SOR_threshold=$5\n",
    "MQ_threshold=$6\n",
    "MQRankSum_threshold=$7\n",
    "ReadPosRankSum_threshold=$8\n",
    "\n",
    "\n",
    "module load java/jdk1.8.0_231\n",
    "module load GATK/4.9.1\n",
    "\n",
    "\n",
    "# 1. Filters to derive biallelic SNP sites where all samples within the regime are called\n",
    "\n",
    "gatk SelectVariants -R '$REF_FASTA' \\\n",
    "-V '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $regime)/$(basename $regime)_SNP_DPfiltered.vcf.gz \\\n",
    "--max-nocall-number 0 \\\n",
    "--restrict-alleles-to BIALLELIC \\\n",
    "-O '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $regime)/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC.vcf\n",
    "\n",
    "# 2. Filters to derive sites that passed the hard-filters\n",
    "\n",
    "gatk VariantFiltration -R '$REF_FASTA' \\\n",
    "-V '$HOME_DIRECTORY'/vcf_handling/bcftools_DP/$(basename $regime)/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC.vcf \\\n",
    "-O '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination).vcf.gz \\\n",
    "--filter \"QD < $QD_threshold\" \\\n",
    "--filter-name \"QD_threshold\" \\\n",
    "--filter \"FS > $FS_threshold\" \\\n",
    "--filter-name \"FS_threshold\" \\\n",
    "--filter \"SOR > $SOR_threshold\" \\\n",
    "--filter-name \"SOR_threshold\" \\\n",
    "--filter \"MQ < $MQ_threshold\" \\\n",
    "--filter-name \"MQ_threshold\" \\\n",
    "--filter \"MQRankSum < $MQRankSum_threshold\" \\\n",
    "--filter-name \"MQRankSum_threshold\" \\\n",
    "--filter \"ReadPosRankSum < $ReadPosRankSum_threshold\" \\\n",
    "--filter-name \"ReadPosRankSum_threshold\" \n",
    "\n",
    "module load bcftools/1.15.1\n",
    "module load htslib/1.15.1\n",
    "\n",
    "bcftools view -i '\"'\"'FILTER=\"PASS\"'\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination).vcf.gz -o '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS.vcf.gz\n",
    "\n",
    "tabix -p vcf '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS.vcf.gz\n",
    "\n",
    "\n",
    "# 3. VariantsToTable\n",
    "\n",
    "gatk VariantsToTable -R '$REF_FASTA' \\\n",
    "-V '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS.vcf.gz \\\n",
    "-O '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS.table \\\n",
    "-F CHROM -F POS -F REF -F ALT -F QD -F QUAL -F DP \\\n",
    "-F MQRankSum -F ReadPosRankSum -F MQ -F FS -F SOR -F AC -F AF -F AN -F HOM-REF -F HET -F HOM-VAR -F NO-CALL -F TRANSITION -F TYPE -F VAR \\\n",
    "-GF GT -GF AD -GF DP -GF GQ \n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/VariantFiltration.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/VariantFiltration_parameter_combinations | while read a b c d e f g h; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/VariantFiltration.sh \"$a\" \"$b\" \"$c\" \"$d\" \"$e\" \"$f\" \"$g\" \"$h\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/VariantFiltration_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/VariantFiltration_submit_jobs\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e979ab26-932d-44d7-94d5-c437fb06867f",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"35\"></a> \n",
    "### Derive heterozygous singletons\n",
    "This section uses the tables generated in the [hard-filtering step](#34) as input and further filters to derive sites where the mutant allele is found in only one of all samples in each regime (i.e. singletons), sites where the mutant sample is heterozygous and the non-mutant samples are homozygous (for the same allele, either reference or alternate), and site where the alleleic ratio of the mutant allele is between 0.25 and 0.75 in the mutant sample, and 0 in the non-mutant samples of the same regime.\n",
    "\n",
    "Specifically, this section contains three scripts:\n",
    "**[Script 1:](#36)** Split the GT and AD columns to facilitate filtering.\n",
    "**[Script 2:](#37)** Derive singletons. Two types of singltons are derived; the mutant sample is heterozygous and the non-mutant samples are homozygous for the reference allele, or the mutant sample is heterozygous and the non-mutant samples are homozygous for the alternate allele.\n",
    "**[Script 3:](#38)** Derive sites where the allelic ratio of the mutant allele is between 0.25 and 0.75 in the mutant sample, and 0 in the non-mutant samples of the same regime."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2150bf2f-6012-4a4c-bde8-106e21a9aa4e",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"36\"></a> \n",
    "#### Split GT and AD columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e15549cb-ac94-417b-8d9d-49dca54283d9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/VariantFiltration\n",
    "JOB_NAME=split_columns\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=3G\n",
    "#SBATCH --time=0-01:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "\n",
    "regime=$1\n",
    "parameter_combination=$2\n",
    "\n",
    "\n",
    "## split GT columns\n",
    "\n",
    "sed '\"'\"'s/|/\\t/g'\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS.table > '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_temp1.table \n",
    "\n",
    "sed '\"'\"'s/\\//\\t/g'\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_temp1.table > '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_temp2.table \n",
    "\n",
    "## split AD columns\n",
    "\n",
    "sed '\"'\"'s/,/\\t/g'\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_temp2.table > '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited.table \n",
    "\n",
    "rm '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_temp1.table\n",
    "rm '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_temp2.table\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/split_columns.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/VariantFiltration_parameter_combinations | while read a b c d e f g h; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/split_columns.sh \"$a\" \"$b\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/split_columns_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/split_columns_submit_jobs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75205aaa-322a-445e-9b42-23757fbf3e15",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"37\"></a> \n",
    "#### Derive singletons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd208916-f4d0-4601-aaa6-407622ca82dc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/VariantFiltration\n",
    "JOB_NAME=singletons\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=3G\n",
    "#SBATCH --time=0-01:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "\n",
    "regime=$1\n",
    "parameter_combination=$2\n",
    "\n",
    "## HOM-REF == 5, HET == 1\n",
    "\n",
    "awk '\"'\"'($16 == 5) && ($17 == 1) && ($18 == 0) {print $0}'\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited.table > '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited_singletons_HOMREF_HET.table\n",
    "\n",
    "## HOM-VAR == 5, HET == 1\n",
    "\n",
    "awk '\"'\"'($16 == 0) && ($17 == 1) && ($18 == 5) {print $0}'\"'\"' '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited.table > '$HOME_DIRECTORY'/vcf_handling/VariantFiltration/$(basename $regime)_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_$(basename $parameter_combination)_PASS_splited_singletons_HOMVAR_HET.table\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/singletons.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/VariantFiltration_parameter_combinations | while read a b c d e f g h; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/singletons.sh \"$a\" \"$b\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/singletons_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/singletons_submit_jobs\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56028f37-b29d-4470-a9a5-870666aced17",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"38\"></a> \n",
    "#### Filter on the allelic ratio of the mutant allele\n",
    "This script performs further filtering to derive sites where the allelic ratio of mutant allele is between 0.25 and 0.75 in the mutant sample, and no mutant allele is found in non-mutant samples of the same regime.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91c7c24b-5a6b-4383-84c5-bb57f4369895",
   "metadata": {},
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_parameter_type_sample1 | while read a b c; do printf \"\n",
    "## for sample1, HOM-REF/HET\n",
    "awk ' (\\$23 == \\$3) && (\\$24 == \\$4) && (\\$29 == \\$3) && (\\$30 == \\$3) && (\\$35 == \\$3) && (\\$36 == \\$3) && (\\$41 == \\$3) && (\\$42 == \\$3) && (\\$47 == \\$3) && (\\$48 == \\$3) && (\\$53 == \\$3) && (\\$54 == \\$3) && (\\$26/(\\$26+\\$25) >= 0.25) && (\\$26/(\\$26+\\$25) <= 0.75) && (\\$32 == 0) && (\\$38 == 0) && (\\$44 == 0) && (\\$50 == 0) && (\\$56 == 0) ' $HOME_DIRECTORY/vcf_handling/VariantFiltration/\"$a\"_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_\"$b\"_PASS_splited_singletons_HOMREF_HET.table > $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$c\"_\"$b\"_HOMREF_HET.table\n",
    "## for sample1, HOM-VAR/HET\n",
    "awk ' (\\$23 == \\$3) && (\\$24 == \\$4) && (\\$29 == \\$4) && (\\$30 == \\$4) && (\\$35 == \\$4) && (\\$36 == \\$4) && (\\$41 == \\$4) && (\\$42 == \\$4) && (\\$47 == \\$4) && (\\$48 == \\$4) && (\\$53 == \\$4) && (\\$54 == \\$4) && (\\$26/(\\$26+\\$25) >= 0.25) && (\\$26/(\\$26+\\$25) <= 0.75) && (\\$31 == 0) && (\\$37 == 0) && (\\$43 == 0) && (\\$49 == 0) && (\\$55 == 0) ' $HOME_DIRECTORY/vcf_handling/VariantFiltration/\"$a\"_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_\"$b\"_PASS_splited_singletons_HOMVAR_HET.table > $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$c\"_\"$b\"_HOMVAR_HET.table\n",
    "\"; done > $HOME_DIRECTORY/scripts/singletons_filtering_sample1\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_parameter_type_sample2 | while read a b c; do printf \"\n",
    "## for sample2, HOM-REF/HET\n",
    "awk ' (\\$23 == \\$3) && (\\$24 == \\$3) && (\\$29 == \\$3) && (\\$30 == \\$4) && (\\$35 == \\$3) && (\\$36 == \\$3) && (\\$41 == \\$3) && (\\$42 == \\$3) && (\\$47 == \\$3) && (\\$48 == \\$3) && (\\$53 == \\$3) && (\\$54 == \\$3) && (\\$32/(\\$32+\\$31) >= 0.25) && (\\$32/(\\$32+\\$31) <= 0.75) && (\\$26 == 0) && (\\$38 == 0) && (\\$44 == 0) && (\\$50 == 0) && (\\$56 == 0) ' $HOME_DIRECTORY/vcf_handling/VariantFiltration/\"$a\"_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_\"$b\"_PASS_splited_singletons_HOMREF_HET.table > $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$c\"_\"$b\"_HOMREF_HET.table\n",
    "## for sample2, HOM-VAR/HET\n",
    "awk ' (\\$23 == \\$4) && (\\$24 == \\$4) && (\\$29 == \\$3) && (\\$30 == \\$4) && (\\$35 == \\$4) && (\\$36 == \\$4) && (\\$41 == \\$4) && (\\$42 == \\$4) && (\\$47 == \\$4) && (\\$48 == \\$4) && (\\$53 == \\$4) && (\\$54 == \\$4) && (\\$32/(\\$32+\\$31) >= 0.25) && (\\$32/(\\$32+\\$31) <= 0.75) && (\\$25 == 0) && (\\$37 == 0) && (\\$43 == 0) && (\\$49 == 0) && (\\$55 == 0) ' $HOME_DIRECTORY/vcf_handling/VariantFiltration/\"$a\"_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_\"$b\"_PASS_splited_singletons_HOMVAR_HET.table > $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$c\"_\"$b\"_HOMVAR_HET.table\n",
    "\"; done > $HOME_DIRECTORY/scripts/singletons_filtering_sample2\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_parameter_type_sample3 | while read a b c; do printf \"\n",
    "## for sample3, HOM-REF/HET\n",
    "awk ' (\\$23 == \\$3) && (\\$24 == \\$3) && (\\$29 == \\$3) && (\\$30 == \\$3) && (\\$35 == \\$3) && (\\$36 == \\$4) && (\\$41 == \\$3) && (\\$42 == \\$3) && (\\$47 == \\$3) && (\\$48 == \\$3) && (\\$53 == \\$3) && (\\$54 == \\$3) && (\\$38/(\\$38+\\$37) >= 0.25) && (\\$38/(\\$38+\\$37) <= 0.75) && (\\$26 == 0) && (\\$32 == 0) && (\\$44 == 0) && (\\$50 == 0) && (\\$56 == 0) ' $HOME_DIRECTORY/vcf_handling/VariantFiltration/\"$a\"_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_\"$b\"_PASS_splited_singletons_HOMREF_HET.table > $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$c\"_\"$b\"_HOMREF_HET.table\n",
    "## for sample3, HOM-VAR/HET\n",
    "awk ' (\\$23 == \\$4) && (\\$24 == \\$4) && (\\$29 == \\$4) && (\\$30 == \\$4) && (\\$35 == \\$3) && (\\$36 == \\$4) && (\\$41 == \\$4) && (\\$42 == \\$4) && (\\$47 == \\$4) && (\\$48 == \\$4) && (\\$53 == \\$4) && (\\$54 == \\$4) && (\\$38/(\\$38+\\$37) >= 0.25) && (\\$38/(\\$38+\\$37) <= 0.75) && (\\$25 == 0) && (\\$31 == 0) && (\\$43 == 0) && (\\$49 == 0) && (\\$55 == 0) ' $HOME_DIRECTORY/vcf_handling/VariantFiltration/\"$a\"_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_\"$b\"_PASS_splited_singletons_HOMVAR_HET.table > $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$c\"_\"$b\"_HOMVAR_HET.table\n",
    "\"; done > $HOME_DIRECTORY/scripts/singletons_filtering_sample3\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_parameter_type_sample4 | while read a b c; do printf \"\n",
    "## for sample4, HOM-REF/HET\n",
    "awk ' (\\$23 == \\$3) && (\\$24 == \\$3) && (\\$29 == \\$3) && (\\$30 == \\$3) && (\\$35 == \\$3) && (\\$36 == \\$3) && (\\$41 == \\$3) && (\\$42 == \\$4) && (\\$47 == \\$3) && (\\$48 == \\$3) && (\\$53 == \\$3) && (\\$54 == \\$3) && (\\$44/(\\$44+\\$43) >= 0.25) && (\\$44/(\\$44+\\$43) <= 0.75) && (\\$26 == 0) && (\\$32 == 0) && (\\$38 == 0) && (\\$50 == 0) && (\\$56 == 0) ' $HOME_DIRECTORY/vcf_handling/VariantFiltration/\"$a\"_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_\"$b\"_PASS_splited_singletons_HOMREF_HET.table > $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$c\"_\"$b\"_HOMREF_HET.table\n",
    "## for sample4, HOM-VAR/HET\n",
    "awk ' (\\$23 == \\$4) && (\\$24 == \\$4) && (\\$29 == \\$4) && (\\$30 == \\$4) && (\\$35 == \\$4) && (\\$36 == \\$4) && (\\$41 == \\$3) && (\\$42 == \\$4) && (\\$47 == \\$4) && (\\$48 == \\$4) && (\\$53 == \\$4) && (\\$54 == \\$4) && (\\$44/(\\$44+\\$43) >= 0.25) && (\\$44/(\\$44+\\$43) <= 0.75) && (\\$25 == 0) && (\\$31 == 0) && (\\$37 == 0) && (\\$49 == 0) && (\\$55 == 0) ' $HOME_DIRECTORY/vcf_handling/VariantFiltration/\"$a\"_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_\"$b\"_PASS_splited_singletons_HOMVAR_HET.table > $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$c\"_\"$b\"_HOMVAR_HET.table\n",
    "\"; done > $HOME_DIRECTORY/scripts/singletons_filtering_sample4\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_parameter_type_sample5 | while read a b c; do printf \"\n",
    "## for sample5, HOM-REF/HET\n",
    "awk ' (\\$23 == \\$3) && (\\$24 == \\$3) && (\\$29 == \\$3) && (\\$30 == \\$3) && (\\$35 == \\$3) && (\\$36 == \\$3) && (\\$41 == \\$3) && (\\$42 == \\$3) && (\\$47 == \\$3) && (\\$48 == \\$4) && (\\$53 == \\$3) && (\\$54 == \\$3) && (\\$50/(\\$49+\\$50) >= 0.25) && (\\$50/(\\$49+\\$50) <= 0.75) && (\\$26 == 0) && (\\$32 == 0) && (\\$38 == 0) && (\\$44 == 0) && (\\$56 == 0) ' $HOME_DIRECTORY/vcf_handling/VariantFiltration/\"$a\"_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_\"$b\"_PASS_splited_singletons_HOMREF_HET.table > $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$c\"_\"$b\"_HOMREF_HET.table\n",
    "## for sample5, HOM-VAR/HET\n",
    "awk ' (\\$23 == \\$4) && (\\$24 == \\$4) && (\\$29 == \\$4) && (\\$30 == \\$4) && (\\$35 == \\$4) && (\\$36 == \\$4) && (\\$41 == \\$4) && (\\$42 == \\$4) && (\\$47 == \\$3) && (\\$48 == \\$4) && (\\$53 == \\$4) && (\\$54 == \\$4) && (\\$50/(\\$49+\\$50) >= 0.25) && (\\$50/(\\$49+\\$50) <= 0.75) && (\\$25 == 0) && (\\$31 == 0) && (\\$37 == 0) && (\\$43 == 0) && (\\$55 == 0) ' $HOME_DIRECTORY/vcf_handling/VariantFiltration/\"$a\"_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_\"$b\"_PASS_splited_singletons_HOMVAR_HET.table > $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$c\"_\"$b\"_HOMVAR_HET.table\n",
    "\"; done > $HOME_DIRECTORY/scripts/singletons_filtering_sample5\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_parameter_type_sample6 | while read a b c; do printf \"\n",
    "## for sample6, HOM-REF/HET\n",
    "awk ' (\\$23 == \\$3) && (\\$24 == \\$3) && (\\$29 == \\$3) && (\\$30 == \\$3) && (\\$35 == \\$3) && (\\$36 == \\$3) && (\\$41 == \\$3) && (\\$42 == \\$3) && (\\$47 == \\$3) && (\\$48 == \\$3) && (\\$53 == \\$3) && (\\$54 == \\$4) && (\\$56/(\\$55+\\$56) >= 0.25) && (\\$56/(\\$55+\\$56) <= 0.75) && (\\$26 == 0) && (\\$32 == 0) && (\\$38 == 0) && (\\$44 == 0) && (\\$50 == 0) ' $HOME_DIRECTORY/vcf_handling/VariantFiltration/\"$a\"_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_\"$b\"_PASS_splited_singletons_HOMREF_HET.table > $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$c\"_\"$b\"_HOMREF_HET.table\n",
    "## for sample6, HOM-VAR/HET\n",
    "awk ' (\\$23 == \\$4) && (\\$24 == \\$4) && (\\$29 == \\$4) && (\\$30 == \\$4) && (\\$35 == \\$4) && (\\$36 == \\$4) && (\\$41 == \\$4) && (\\$42 == \\$4) && (\\$47 == \\$4) && (\\$48 == \\$4) && (\\$53 == \\$3) && (\\$54 == \\$4) && (\\$56/(\\$55+\\$56) >= 0.25) && (\\$56/(\\$55+\\$56) <= 0.75) && (\\$25 == 0) && (\\$31 == 0) && (\\$37 == 0) && (\\$43 == 0) && (\\$49 == 0) ' $HOME_DIRECTORY/vcf_handling/VariantFiltration/\"$a\"_SNP_DPfiltered_AllCalled_BIALLELIC_VariantFiltration_\"$b\"_PASS_splited_singletons_HOMVAR_HET.table > $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$c\"_\"$b\"_HOMVAR_HET.table\n",
    "\"; done > $HOME_DIRECTORY/scripts/singletons_filtering_sample6\n",
    "\n",
    "#sh -e $HOME_DIRECTORY/scripts/singletons_filtering_sample1\n",
    "#sh -e $HOME_DIRECTORY/scripts/singletons_filtering_sample2\n",
    "#sh -e $HOME_DIRECTORY/scripts/singletons_filtering_sample3\n",
    "#sh -e $HOME_DIRECTORY/scripts/singletons_filtering_sample4\n",
    "#sh -e $HOME_DIRECTORY/scripts/singletons_filtering_sample5\n",
    "#sh -e $HOME_DIRECTORY/scripts/singletons_filtering_sample6"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da4ced50-821c-4f69-af4f-754b05f8fd33",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id=\"39\"></a> \n",
    "### Derive the number of candidate germline mutations for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5666c1bf-3bde-43ec-adc9-b58a8e9e4cd0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "\n",
    "touch $HOME_DIRECTORY/de_novo_mutation/consensus_approach_number_GM_before_manual_inspection\n",
    "\n",
    "less $HOME_DIRECTORY/regime_parameter_type_sample_wide | while read a b c d e f g h i; do printf \"\n",
    "wc -l $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$d\"_\"$b\"_\"$c\".table >> $HOME_DIRECTORY/de_novo_mutation/consensus_approach_number_GM_before_manual_inspection\n",
    "wc -l $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$e\"_\"$b\"_\"$c\".table >> $HOME_DIRECTORY/de_novo_mutation/consensus_approach_number_GM_before_manual_inspection\n",
    "wc -l $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$f\"_\"$b\"_\"$c\".table >> $HOME_DIRECTORY/de_novo_mutation/consensus_approach_number_GM_before_manual_inspection\n",
    "wc -l $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$g\"_\"$b\"_\"$c\".table >> $HOME_DIRECTORY/de_novo_mutation/consensus_approach_number_GM_before_manual_inspection\n",
    "wc -l $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$h\"_\"$b\"_\"$c\".table >> $HOME_DIRECTORY/de_novo_mutation/consensus_approach_number_GM_before_manual_inspection\n",
    "wc -l $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$i\"_\"$b\"_\"$c\".table >> $HOME_DIRECTORY/de_novo_mutation/consensus_approach_number_GM_before_manual_inspection\n",
    "\"; done > $HOME_DIRECTORY/scripts/HaplotypeCaller_number_GM_before_manual_inspection\n",
    "\n",
    "sh -e $HOME_DIRECTORY/scripts/HaplotypeCaller_number_GM_before_manual_inspection\n",
    "\n",
    "sed -i 's/\\/gpfs\\/home\\/wwh23wcu\\/de_novo_mutation\\/consensus_approach\\///' $HOME_DIRECTORY/de_novo_mutation/consensus_approach_number_GM_before_manual_inspection\n",
    "\n",
    "sed -i 's/_/ /' $HOME_DIRECTORY/de_novo_mutation/consensus_approach_number_GM_before_manual_inspection\n",
    "sed -i 's/_/ /' $HOME_DIRECTORY/de_novo_mutation/consensus_approach_number_GM_before_manual_inspection\n",
    "\n",
    "sed -i 's/.table//' $HOME_DIRECTORY/de_novo_mutation/consensus_approach_number_GM_before_manual_inspection\n",
    "\n",
    "\n",
    "awk -F' ' 'BEGIN { SUBSEP = OFS = FS } { s[$2,$3] += $1; ++c[$2,$3] } END { for (i in s) { print i, s[i], c[i] } }' $HOME_DIRECTORY/de_novo_mutation/consensus_approach_number_GM_before_manual_inspection | sort | uniq > $HOME_DIRECTORY/de_novo_mutation/consensus_approach_number_GM_before_manual_inspection_temp\n",
    "\n",
    "mv $HOME_DIRECTORY/de_novo_mutation/consensus_approach_number_GM_before_manual_inspection_temp $HOME_DIRECTORY/de_novo_mutation/consensus_approach_number_GM_before_manual_inspection\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b33bca6-7f14-4340-88b0-1d4ac46e1eaf",
   "metadata": {},
   "source": [
    "<a id=\"40\"></a> \n",
    "### Derive bed files for manual inspection\n",
    "This script derives a bed file for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f3ee521-9579-4e1d-9716-d1f9a9ae25d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_sample | while read a b; do printf \"\n",
    "awk '{print \\$1\\\" \\\"\\$2\\\" \\\"\\$3\\\" \\\"\\$4}' $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$b\"_1_HOMREF_HET.table > $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$b\"_1_HOMREF_HET.bed\n",
    "awk '{print \\$1\\\" \\\"\\$2\\\" \\\"\\$4\\\" \\\"\\$3}' $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$b\"_1_HOMVAR_HET.table > $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$b\"_1_HOMVAR_HET.bed\n",
    "\n",
    "cat $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$b\"_1_HOMREF_HET.bed $HOME_DIRECTORY/de_novo_mutation/consensus_approach/\"$b\"_1_HOMVAR_HET.bed > $HOME_DIRECTORY/de_novo_mutation/bed/\"$b\"_consensus_bed\n",
    "\"; done > $HOME_DIRECTORY/scripts/consensus_derive_bed\n",
    "\n",
    "#sh -e $HOME_DIRECTORY/scripts/consensus_derive_bed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d7296d1-fbeb-474f-9d94-64cd813966fb",
   "metadata": {},
   "source": [
    "<a id=\"41\"></a> \n",
    "## Callability"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbdcca41-ffac-4ddc-a801-30a81138dfae",
   "metadata": {},
   "source": [
    "<a id=\"46\"></a> \n",
    "### Derive genotyped sites for each regime\n",
    "This script filters out sites with missing genotype in any of the samples of the same regime and keeps sites that all samples of the same regime are successfully genotyped."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "607b3aeb-9ad3-4623-97d7-2bed41002081",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/callability\n",
    "JOB_NAME=callability_cohort\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=3G\n",
    "#SBATCH --time=0-04:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "\n",
    "module load bcftools/1.15.1\n",
    "\n",
    "bcftools view -i '\"'\"' FORMAT/GT[0] != \"mis\" && FORMAT/GT[1] != \"mis\" && FORMAT/GT[2] != \"mis\"  && FORMAT/GT[3] != \"mis\" && FORMAT/GT[4] != \"mis\" && FORMAT/GT[5] != \"mis\" '\"'\"' '$HOME_DIRECTORY'/vcf_handling/GenotypeGVCFs/$(basename $regime)_GenotypeGVCFs.g.vcf.gz \\\n",
    "--output '$HOME_DIRECTORY'/vcf_handling/callability/$(basename $regime)_GenotypeGVCFs_callable.g.vcf.gz\n",
    "\n",
    "bcftools index -t '$HOME_DIRECTORY'/vcf_handling/callability/$(basename $regime)_GenotypeGVCFs_callable.g.vcf.gz\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/callability_cohort.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_list | while read a b; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/callability_cohort.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/callability_cohort_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/callability_cohort_submit_jobs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b2abdbe-096f-4c3e-a306-97b528fd2d9f",
   "metadata": {},
   "source": [
    "<a id=\"43\"></a> \n",
    "### Derive the number of callable sites for each regime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d47f3fe-3794-4f42-ba09-8d0b8bbb5fe3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "\n",
    "touch $HOME_DIRECTORY/de_novo_mutation/consensus_approach_callability_regime\n",
    "\n",
    "less $HOME_DIRECTORY/regime_list | while read a; do printf \"\n",
    "bcftools index --nrecords $HOME_DIRECTORY/vcf_handling/callability/\"$a\"_GenotypeGVCFs_callable.g.vcf.gz >> $HOME_DIRECTORY/de_novo_mutation/consensus_approach_callability_regime\n",
    "\"; done > $HOME_DIRECTORY/scripts/callability_cohort_derive_results\n",
    "\n",
    "\n",
    "sed -i '1i module load bcftools/1.15.1' $HOME_DIRECTORY/scripts/callability_cohort_derive_results\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/callability_cohort_derive_results\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12b7d0cf-2ca4-47cc-a05d-2a5b7ebb594b",
   "metadata": {},
   "source": [
    "<a id=\"44\"></a> \n",
    "# Variant calling by the probabilistic approach"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "610d74dd-5139-49e8-8ba9-46ebab07cca5",
   "metadata": {},
   "source": [
    "<a id=\"45\"></a> \n",
    "## Merge bam files for each regime\n",
    "This script uses the picard MergeSamFiles tool to merge single-sample bam files of the same regime into a multi-sample bam file. Total three multi-sample bam files, one for each regime, are generated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f751b92e-1361-4f76-abf7-48b4ef270db0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output\n",
    "JOB_NAME=picard_MergeSamFiles\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=60G\n",
    "#SBATCH --time=0-48:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "regime_sample1=$2\n",
    "regime_sample2=$3\n",
    "regime_sample3=$4\n",
    "regime_sample4=$5\n",
    "regime_sample5=$6\n",
    "regime_sample6=$7\n",
    "\n",
    "module load java/jdk1.8.0_231\n",
    "module load picard/2.24.1\n",
    "\n",
    "\n",
    "java -jar /gpfs/software/ada/picard/2.24.1/picard.jar MergeSamFiles \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $regime_sample1)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $regime_sample2)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $regime_sample3)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $regime_sample4)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $regime_sample5)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $regime_sample6)_viewed_marked_addg.bam \\\n",
    "O='$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles.bam\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/MergeSamFiles.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_sample_wide | while read a b c d e f g; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/MergeSamFiles.sh \"$a\" \"$b\" \"$c\" \"$d\" \"$e\" \"$f\" \"$g\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/MergeSamFiles_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/MergeSamFiles_submit_jobs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71533bdd-5be4-410c-a26b-34b7eaab0383",
   "metadata": {},
   "source": [
    "<a id=\"46\"></a> \n",
    "## Add a readgroup for the ancestor to the header of the merged bam\n",
    "This script uses the samtools view tool to add a dummy readgroup for the ancestor to the merged, multi-sample bam file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da78620d-6654-403d-8218-3022dc20cc65",
   "metadata": {},
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/accuMUlate\n",
    "JOB_NAME=addg\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=12G\n",
    "#SBATCH --time=0-12:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "\n",
    "\n",
    "module load samtools/1.16.1\n",
    "\n",
    "samtools view -H '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles.bam > '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles.head.sam\n",
    "\n",
    "sed '\"'\"'19 i @RG\\tID:Ancestral\\tLB:lib1 PL:illumina_TruSeq\\tSM:Ancestral\\tPU:unit1'\"'\"' '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles.head.sam > '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_addg.head.sam\n",
    "\n",
    "samtools view '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles.bam | cat '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_addg.head.sam - | samtools view -Sb - > '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_addg.bam\n",
    "\n",
    "samtools index '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_addg.bam\n",
    "\n",
    "\n",
    "module load python/3.10.0\n",
    "\n",
    "samtools view -H '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_addg.bam | python '$HOME_DIRECTORY'/extract_samples.py Ancestral - >> '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_params.ini\n",
    "\n",
    "\n",
    "cat '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_params.ini '$HOME_DIRECTORY'/accuMUlate_parameters > '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_params_temp.ini && mv '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_params_temp.ini '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_params.ini\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/addg.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_list | while read a; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/addg.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/addg_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/addg_submit_jobs\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f70f731e-c006-4d28-b7a2-5d665c1b989d",
   "metadata": {},
   "source": [
    "<a id=\"47\"></a> \n",
    "## Call variants\n",
    "This script uses the accuMUlate tool to call variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88821155-6f2c-4c1c-befc-38aa89a29590",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "REF_FASTA_DECOMPRESSED=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/accuMUlate\n",
    "JOB_NAME=accuMUlate\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=30G\n",
    "#SBATCH --time=5-00:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "\n",
    "module load accuMUlate/0.2.2\n",
    "\n",
    "/gpfs/software/ada/accuMUlate/0.2.2/accuMUlate \\\n",
    "-b '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_addg.bam \\\n",
    "-r '$REF_FASTA_DECOMPRESSED' \\\n",
    "-q 13 \\\n",
    "-m 13 \\\n",
    "-p 0.10000000000000001 \\\n",
    "--ploidy-ancestor 2 \\\n",
    "--ploidy-descendant 2 \\\n",
    "-c '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_params.ini \\\n",
    "-o '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_putative_mutations.tsv\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/accuMUlate.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_list | while read a b c d e f g; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/accuMUlate.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/accuMUlate_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/accuMUlate_submit_jobs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47e583ba-29e0-40f2-8470-fee1b9ab16e7",
   "metadata": {},
   "source": [
    "<a id=\"48\"></a> \n",
    "## Variant filtering\n",
    "This scripts filters to keep the following:\n",
    "1. Sites outside of soft-masked regions;\n",
    "2. Sites where estimated probability of at least one mutation at this site = 1;\n",
    "3. Sites where estimated probability of exactly one mutation at this site = 1; \n",
    "4. Sites where probability of the mutation direction above being correct = 1; \n",
    "5. Homozygous to heterozygous mutations;\n",
    "6. Sites where the number of reads for the mutant allele in non-mutant samples = 0; \n",
    "7. Sites where the allelic ratio for the mutant allele in the mutant sample is between 0.25 and 0.75;\n",
    "8. Sites where the AD test statistic for mapping quality difference between mutant and non-mutant alleles =< 1.95;\n",
    "9. Sites where the AD test statistic for insert size differfence between inferred insert size between mutant and non-mutant alleles =< 1.95;\n",
    "10. Sites where the p-value from Fisher's exact test of strand bias between mutant and non-mutant alleles is > 0.05;\n",
    "11. Sites where the p-value from Fisher's exact test of pair-mapping rate difference between mutant and non-mutant alleles is > 0.05."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc8d0452-cd11-497a-b187-29ce1551defa",
   "metadata": {},
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "less $HOME_DIRECTORY/regime_list | while read a; do printf \"\n",
    "\n",
    "awk '\"' $4~/[A-Z]/ '\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp1.tsv\n",
    "\n",
    "awk '\"' $7==1 && $8==1 && $9==1 {print $0}'\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp1.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp2.tsv\n",
    "\n",
    "awk '\"' { if (substr($6,1,1) == substr($6,2,1) && substr($6,1,1) == substr($6,5,1) && substr($6,5,1) != substr($6,6,1) ) print $0 }'\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp2.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp3.tsv\n",
    "\n",
    "awk '\"' $15==0 {print $0}'\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp3.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp4.tsv\n",
    "\n",
    "awk '\"' $14/($14+$12+$13) >=0.25 && $14/($14+$12+$13) <= 0.75 {print $0}'\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp4.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp5.tsv\n",
    "\n",
    "awk '\"' $16 <= 1.95 {print $0} '\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp5.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp6.tsv\n",
    "\n",
    "awk '\"' $17 <= 1.95 {print $0} '\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp6.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp7.tsv\n",
    "\n",
    "awk '\"' $18 > 0.05 {print $0} '\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp7.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp8.tsv\n",
    "\n",
    "awk '\"' $19 > 0.05 {print $0} '\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp8.tsv > $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach/\"$a\"_filtered_mutations.tsv\n",
    "\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp1.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp2.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp3.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp4.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp5.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp6.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp7.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp8.tsv\n",
    "\"; done > $HOME_DIRECTORY/scripts/accuMUlate_filtering\n",
    "\n",
    "#sh -e $HOME_DIRECTORY/scripts/accuMUlate_filtering\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "416225b4-2f22-4ae4-b5af-b04aab3cc09c",
   "metadata": {},
   "source": [
    "<a id=\"49\"></a> \n",
    "## Derive the number of candidate germline mutations for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "415db7be-e9b0-4922-8e30-0d21851e8932",
   "metadata": {},
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "touch $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach_number_GM\n",
    "\n",
    "less $HOME_DIRECTORY/regime_sample | while read a b; do printf \"\n",
    "grep '\"$b\"' $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach/\"$a\"_filtered_mutations.tsv > $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach/\"$b\"_filtered_mutations.tsv\n",
    "wc -l $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach/\"$b\"_filtered_mutations.tsv >> $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach_number_GM_before_manual_inspection\n",
    "\"; done > $HOME_DIRECTORY/scripts/accuMUlate_number_GM_before_manual_inspection\n",
    "\n",
    "sh -e $HOME_DIRECTORY/scripts/accuMUlate_number_GM_before_manual_inspection\n",
    "\n",
    "sed -i 's/\\/gpfs\\/home\\/wwh23wcu\\/de_novo_mutation\\/probabilistic_approach\\///' $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach_number_GM_before_manual_inspection\n",
    "\n",
    "sed -i 's/_filtered_mutations.tsv//' $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach_number_GM_before_manual_inspection\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1aa3b290-e6e2-44c0-a928-3de2840a4fe6",
   "metadata": {},
   "source": [
    "<a id=\"50\"></a> \n",
    "## Derive denominator\n",
    "This script uses the accuMUlate denominate tool to derive the denominator for mutation rate calculation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86e5a443-dd10-4bd8-b4ed-6ec17e0a8bf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Derive denominator\n",
    "\n",
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "REF_FASTA_DECOMPRESSED=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/accuMUlate\n",
    "JOB_NAME=denominator\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=30G\n",
    "#SBATCH --time=5-00:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "\n",
    "module load accuMUlate/0.2.2\n",
    "\n",
    "accuMUlate denominate -c '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_params.ini -r '$REF_FASTA_DECOMPRESSED' -b '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_addg.bam > '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_denom.out\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/denominator.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_list | while read a; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/denominator.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/denominator_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/denominator_submit_jobs\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f500c3bb-2e1c-4b9e-a81e-c2d7290951d1",
   "metadata": {},
   "source": [
    "<a id=\"51\"></a> \n",
    "## Derive bed files for manual inspection\n",
    "This script derives a bed file for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95d0ca2c-5b99-46e4-aec8-cb4f587e759b",
   "metadata": {},
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "less $HOME_DIRECTORY/regime_sample | while read a b; do echo \"\n",
    "\n",
    "awk '{print \\$1\\\" \\\"\\$3\\\" \\\"\\$4\\\" \\\"substr(\\$6,6,1)}' $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach/\"$b\"_filtered_mutations.tsv > $HOME_DIRECTORY/de_novo_mutation/bed/\"$b\"_probabilistic_bed\n",
    "\n",
    "\"; done > $HOME_DIRECTORY/scripts/probabilistic_derive_bed\n",
    "\n",
    "sh -e $HOME_DIRECTORY/scripts/probabilistic_derive_bed\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33b4671d-d8ea-4910-957e-da379f4bd678",
   "metadata": {},
   "source": [
    "<a id=\"52\"></a> \n",
    "# Manual inspection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eac620c4-0783-4fc0-8be7-2c2b81a8fe95",
   "metadata": {},
   "source": [
    "<a id=\"53\"></a> \n",
    "## Merge and annotate the bed files from both approaches for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "946ed13f-1de5-4402-8adf-be51748c63b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "less $HOME_DIRECTORY/regime_sample | while read a b; do echo \"\n",
    "awk '\\$5 = \\\"consensus\\\" {print \\$1\\\" \\\"\\$2\\\" \\\"\\$3\\\" \\\"\\$4\\\" \\\"\\$5}' $HOME_DIRECTORY/de_novo_mutation/bed/\"$b\"_consensus_bed > $HOME_DIRECTORY/de_novo_mutation/bed/\"$b\"_consensus_bed_temp\n",
    "awk '\\$5 = \\\"probabilistic\\\" {print \\$1\\\" \\\"\\$2\\\" \\\"\\$3\\\" \\\"\\$4\\\" \\\"\\$5}' $HOME_DIRECTORY/de_novo_mutation/bed/\"$b\"_probabilistic_bed > $HOME_DIRECTORY/de_novo_mutation/bed/\"$b\"_probabilistic_bed_temp\n",
    "\n",
    "cat $HOME_DIRECTORY/de_novo_mutation/bed/\"$b\"_consensus_bed_temp $HOME_DIRECTORY/de_novo_mutation/bed/\"$b\"_probabilistic_bed_temp > $HOME_DIRECTORY/de_novo_mutation/bed/\"$b\"_bed_temp\n",
    "\n",
    "sort $HOME_DIRECTORY/de_novo_mutation/bed/\"$b\"_bed_temp > $HOME_DIRECTORY/de_novo_mutation/bed/manual_inspection/\"$b\"_bed_before_manual_inspection\n",
    "\n",
    "rm $HOME_DIRECTORY/de_novo_mutation/bed/*_temp\n",
    "\"; done > $HOME_DIRECTORY/scripts/prepare_bed_files\n",
    "\n",
    "sh -e $HOME_DIRECTORY/scripts/prepare_bed_files\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b59072a-894f-4318-a3fe-ce62d6aaea47",
   "metadata": {},
   "source": [
    "<a id=\"54\"></a> \n",
    "## Derive single-sample vcf\n",
    "This script splits the multi-sample vcf derived from joint genotyping to six single-sample vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6950a7e7-65bf-4d6c-b12f-3298c3d4541c",
   "metadata": {},
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/bcftools_split\n",
    "JOB_NAME=bcftools_split\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=3G\n",
    "#SBATCH --time=0-04:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "\n",
    "module load bcftools/1.15.1\n",
    "module load htslib/1.15.1\n",
    "\n",
    "bcftools +split '$HOME_DIRECTORY'/vcf_handling/GenotypeGVCFs/$(basename $regime)_GenotypeGVCFs.g.vcf.gz -O z -o '$HOME_DIRECTORY'/de_novo_mutation/bed/manual_inspection/\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/bcftools_split.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_list | while read a; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/bcftools_split.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/bcftools_split_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/bcftools_split_submit_jobs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "580d2232-0911-4db0-8837-e49dac512bfd",
   "metadata": {},
   "source": [
    "<a id=\"55\"></a> \n",
    "## Index single-sample vcf\n",
    "This script indexes the single-sample vcf files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40b61f26-82fd-4704-8c30-db4ed2431569",
   "metadata": {},
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/bcftools_split\n",
    "JOB_NAME=bcftools_split_tabix\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=3G\n",
    "#SBATCH --time=0-04:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "sample_ID=$2\n",
    "\n",
    "module load bcftools/1.15.1\n",
    "module load htslib/1.15.1\n",
    "\n",
    "tabix -p vcf '$HOME_DIRECTORY'/de_novo_mutation/bed/manual_inspection/$(basename $sample_ID).vcf.gz\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/bcftools_split_tabix.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_sample_field_for_DP | while read a b c; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/bcftools_split_tabix.sh \"$a\" \"$b\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/bcftools_split_tabix_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/bcftools_split_tabix_submit_jobs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48c1dbb6-c667-4166-9ef6-3931342f0385",
   "metadata": {},
   "source": [
    "<a id=\"56\"></a> \n",
    "## Files needed for manual inspection\n",
    "\n",
    "Files needed for manual inspection include:\n",
    "1 and 2: the reference assembly;\n",
    "3 and 4: bam and bai files for each sample;\n",
    "5 and 6: single-sample vcf and its index for each sample;\n",
    "7: bed files containing the coordination for each candidate germline mutation for each sample.\n",
    "\n",
    "1. /gpfs/scratch/wwh23wcu/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna\n",
    "2. /gpfs/scratch/wwh23wcu/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.fai\n",
    "3. /gpfs/scratch/wwh23wcu/bam_files/*_viewed_marked_addg.bam \n",
    "4. /gpfs/scratch/wwh23wcu/bam_files/*_viewed_marked_addg.bam.bai\n",
    "5. /gpfs/home/wwh23wcu/de_novo_mutation/bed/manual_inspection/*.vcf.gz\n",
    "6. /gpfs/home/wwh23wcu/de_novo_mutation/bed/manual_inspection/*.vcf.gz.tbi\n",
    "7. /gpfs/home/wwh23wcu/de_novo_mutation/bed/manual_inspection/*_bed_before_manual_inspection\n",
    "\n",
    "For each sample, load the reference assembly and the vcf in IGV; then load the bam files for all samples of the same regime in IGV. Manually inspect individual candidate germline mutation and annotate 'ACCEPTED' or 'REJECTED' in the bed file. When finished, save the bed file as sampleID_bed_after_manual_inspection, and upload it to /gpfs/home/wwh23wcu/de_novo_mutation/bed/manual_inspection/."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27eff685-8550-4426-8e0e-6ce6acd63b40",
   "metadata": {},
   "source": [
    "<a id=\"57\"></a> \n",
    "# Find overlapping variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41554c43-8e96-4816-809a-abc59aa902a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "less $HOME_DIRECTORY/regime_sample | while read a b; do echo \"\n",
    "awk '{print \\$1\\\" \\\"\\$2\\\" \\\"\\$3\\\" \\\"\\$4}' $HOME_DIRECTORY/de_novo_mutation/bed/manual_inspection/\"$b\"_bed_after_manual_inspection | sort | uniq -d > $HOME_DIRECTORY/de_novo_mutation/\"$b\"_bed_after_manual_inspection_overlapping\n",
    "\"; done > $HOME_DIRECTORY/scripts/find_overlapping_variants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30760c91-53e8-4c0d-b04e-4a6d2f68054b",
   "metadata": {},
   "source": [
    "<a id=\"58\"></a> \n",
    "# Aggregate the results\n",
    "This script aggregates callability and the number of accepted germline mutations for each sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7227b6ca-f03e-4aea-a5c3-bb89809d78b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "paste -d ' ' $HOME_DIRECTORY/regime_list $HOME_DIRECTORY/de_novo_mutation/consensus_approach_callability_regime > $HOME_DIRECTORY/de_novo_mutation/consensus_approach_callability_regime_temp && mv $HOME_DIRECTORY/de_novo_mutation/consensus_approach_callability_regime_temp $HOME_DIRECTORY/de_novo_mutation/consensus_approach_callability_regime\n",
    "\n",
    "perl -ne 'print $_ x 36' $HOME_DIRECTORY/de_novo_mutation/consensus_approach_callability_regime > $HOME_DIRECTORY/de_novo_mutation/consensus_approach_callability_regime_temp\n",
    "\n",
    "paste -d ' ' $HOME_DIRECTORY/de_novo_mutation/consensus_approach_callability_regime_temp $HOME_DIRECTORY/de_novo_mutation/consensus_approach_number_GM > $HOME_DIRECTORY/de_novo_mutation/consensus_approach_results_temp\n",
    "\n",
    "awk '{$6=\"\"; print $0}' $HOME_DIRECTORY/de_novo_mutation/consensus_approach_results_temp > $HOME_DIRECTORY/de_novo_mutation/consensus_approach_results\n",
    "\n",
    "\n",
    "rm $HOME_DIRECTORY/de_novo_mutation/consensus_approach_callability_regime_temp\n",
    "rm $HOME_DIRECTORY/de_novo_mutation/consensus_approach_results_temp"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "toc-autonumbering": true,
  "toc-showcode": true,
  "toc-showmarkdowntxt": false,
  "toc-showtags": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
