{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c7779719-ec6c-40a8-8ffd-3e0845e04ba5",
   "metadata": {},
   "source": [
    "<a id=\"44\"></a> \n",
    "# Variant calling by the probabilistic approach"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9f78d31-06fa-41cf-b1e2-c7ca964d8129",
   "metadata": {},
   "source": [
    "<a id=\"44\"></a> \n",
    "# Variant calling by the probabilistic approach"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e338cfc7-283d-4f2b-9868-57a95cb85b40",
   "metadata": {},
   "source": [
    "<a id=\"45\"></a> \n",
    "## Merge bam files for each regime\n",
    "This script uses the picard MergeSamFiles tool to merge single-sample bam files of the same regime into a multi-sample bam file. Total three multi-sample bam files, one for each regime, are generated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32659e16-25d7-4934-8cda-709c40b1a544",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output\n",
    "JOB_NAME=picard_MergeSamFiles\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=60G\n",
    "#SBATCH --time=0-48:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "regime_sample1=$2\n",
    "regime_sample2=$3\n",
    "regime_sample3=$4\n",
    "regime_sample4=$5\n",
    "regime_sample5=$6\n",
    "regime_sample6=$7\n",
    "\n",
    "module load java/jdk1.8.0_231\n",
    "module load picard/2.24.1\n",
    "\n",
    "\n",
    "java -jar /gpfs/software/ada/picard/2.24.1/picard.jar MergeSamFiles \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $regime_sample1)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $regime_sample2)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $regime_sample3)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $regime_sample4)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $regime_sample5)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $regime_sample6)_viewed_marked_addg.bam \\\n",
    "O='$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles.bam\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/MergeSamFiles.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_sample_wide | while read a b c d e f g; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/MergeSamFiles.sh \"$a\" \"$b\" \"$c\" \"$d\" \"$e\" \"$f\" \"$g\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/MergeSamFiles_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/MergeSamFiles_submit_jobs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4460de5c-1220-4403-9d81-ae2cbcc8912a",
   "metadata": {},
   "source": [
    "<a id=\"46\"></a> \n",
    "## Add a readgroup for the ancestor to the header of the merged bam\n",
    "This script uses the samtools view tool to add a dummy readgroup for the ancestor to the merged, multi-sample bam file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f1b81d6-6564-473b-884c-23c01e757f6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/accuMUlate\n",
    "JOB_NAME=addg\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=12G\n",
    "#SBATCH --time=0-12:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "\n",
    "\n",
    "module load samtools/1.16.1\n",
    "\n",
    "samtools view -H '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles.bam > '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles.head.sam\n",
    "\n",
    "sed '\"'\"'19 i @RG\\tID:Ancestral\\tLB:lib1 PL:illumina_TruSeq\\tSM:Ancestral\\tPU:unit1'\"'\"' '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles.head.sam > '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_addg.head.sam\n",
    "\n",
    "samtools view '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles.bam | cat '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_addg.head.sam - | samtools view -Sb - > '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_addg.bam\n",
    "\n",
    "samtools index '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_addg.bam\n",
    "\n",
    "\n",
    "module load python/3.10.0\n",
    "\n",
    "samtools view -H '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_addg.bam | python '$HOME_DIRECTORY'/extract_samples.py Ancestral - >> '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_params.ini\n",
    "\n",
    "\n",
    "cat '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_params.ini '$HOME_DIRECTORY'/accuMUlate_parameters > '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_params_temp.ini && mv '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_params_temp.ini '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_params.ini\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/addg.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_list | while read a; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/addg.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/addg_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/addg_submit_jobs\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2d5b313-153f-4e7b-9bb1-1bbbad5351e8",
   "metadata": {},
   "source": [
    "<a id=\"47\"></a> \n",
    "## Call variants\n",
    "This script uses the accuMUlate tool to call variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d8d5493-23dd-47a4-abb1-358913d2c316",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "REF_FASTA_DECOMPRESSED=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/accuMUlate\n",
    "JOB_NAME=accuMUlate\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=30G\n",
    "#SBATCH --time=5-00:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "\n",
    "module load accuMUlate/0.2.2\n",
    "\n",
    "/gpfs/software/ada/accuMUlate/0.2.2/accuMUlate \\\n",
    "-b '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_addg.bam \\\n",
    "-r '$REF_FASTA_DECOMPRESSED' \\\n",
    "-q 13 \\\n",
    "-m 13 \\\n",
    "-p 0.10000000000000001 \\\n",
    "--ploidy-ancestor 2 \\\n",
    "--ploidy-descendant 2 \\\n",
    "-c '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_params.ini \\\n",
    "-o '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_putative_mutations.tsv\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/accuMUlate.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_list | while read a b c d e f g; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/accuMUlate.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/accuMUlate_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/accuMUlate_submit_jobs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2ac6c6e-2e99-419f-8d9b-af5d759d5bb9",
   "metadata": {},
   "source": [
    "<a id=\"48\"></a> \n",
    "## Variant filtering\n",
    "This scripts filters to keep the following:\n",
    "1. Sites outside of soft-masked regions;\n",
    "2. Sites where estimated probability of at least one mutation at this site = 1;\n",
    "3. Sites where estimated probability of exactly one mutation at this site = 1; \n",
    "4. Sites where probability of the mutation direction above being correct = 1; \n",
    "5. Homozygous to heterozygous mutations;\n",
    "6. Sites where the number of reads for the mutant allele in non-mutant samples = 0; \n",
    "7. Sites where the allelic ratio for the mutant allele in the mutant sample is between 0.25 and 0.75;\n",
    "8. Sites where the AD test statistic for mapping quality difference between mutant and non-mutant alleles =< 1.95;\n",
    "9. Sites where the AD test statistic for insert size differfence between inferred insert size between mutant and non-mutant alleles =< 1.95;\n",
    "10. Sites where the p-value from Fisher's exact test of strand bias between mutant and non-mutant alleles is > 0.05;\n",
    "11. Sites where the p-value from Fisher's exact test of pair-mapping rate difference between mutant and non-mutant alleles is > 0.05."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cabef9d6-e44d-45f2-98a7-4a0d91f53810",
   "metadata": {},
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "less $HOME_DIRECTORY/regime_list | while read a; do printf \"\n",
    "\n",
    "awk '\"' $4~/[A-Z]/ '\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp1.tsv\n",
    "\n",
    "awk '\"' $7==1 && $8==1 && $9==1 {print $0}'\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp1.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp2.tsv\n",
    "\n",
    "awk '\"' { if (substr($6,1,1) == substr($6,2,1) && substr($6,1,1) == substr($6,5,1) && substr($6,5,1) != substr($6,6,1) ) print $0 }'\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp2.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp3.tsv\n",
    "\n",
    "awk '\"' $15==0 {print $0}'\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp3.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp4.tsv\n",
    "\n",
    "awk '\"' $14/($14+$12+$13) >=0.25 && $14/($14+$12+$13) <= 0.75 {print $0}'\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp4.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp5.tsv\n",
    "\n",
    "awk '\"' $16 <= 1.95 {print $0} '\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp5.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp6.tsv\n",
    "\n",
    "awk '\"' $17 <= 1.95 {print $0} '\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp6.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp7.tsv\n",
    "\n",
    "awk '\"' $18 > 0.05 {print $0} '\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp7.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp8.tsv\n",
    "\n",
    "awk '\"' $19 > 0.05 {print $0} '\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp8.tsv > $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach/\"$a\"_filtered_mutations.tsv\n",
    "\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp1.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp2.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp3.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp4.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp5.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp6.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp7.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp8.tsv\n",
    "\"; done > $HOME_DIRECTORY/scripts/accuMUlate_filtering\n",
    "\n",
    "#sh -e $HOME_DIRECTORY/scripts/accuMUlate_filtering\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e617e161-56b4-4fd3-ad9d-e3d1f93b87d7",
   "metadata": {},
   "source": [
    "<a id=\"49\"></a> \n",
    "## Derive the number of candidate germline mutations for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61fb45ce-e833-40d4-b737-50a5a37473e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "touch $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach_number_GM\n",
    "\n",
    "less $HOME_DIRECTORY/regime_sample | while read a b; do printf \"\n",
    "grep '\"$b\"' $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach/\"$a\"_filtered_mutations.tsv > $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach/\"$b\"_filtered_mutations.tsv\n",
    "wc -l $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach/\"$b\"_filtered_mutations.tsv >> $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach_number_GM_before_manual_inspection\n",
    "\"; done > $HOME_DIRECTORY/scripts/accuMUlate_number_GM_before_manual_inspection\n",
    "\n",
    "sh -e $HOME_DIRECTORY/scripts/accuMUlate_number_GM_before_manual_inspection\n",
    "\n",
    "sed -i 's/\\/gpfs\\/home\\/wwh23wcu\\/de_novo_mutation\\/probabilistic_approach\\///' $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach_number_GM_before_manual_inspection\n",
    "\n",
    "sed -i 's/_filtered_mutations.tsv//' $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach_number_GM_before_manual_inspection\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2301b8a2-5b30-4a03-b579-e152bcbb00b8",
   "metadata": {},
   "source": [
    "<a id=\"45\"></a> \n",
    "## Merge bam files for each regime\n",
    "This script uses the picard MergeSamFiles tool to merge single-sample bam files of the same regime into a multi-sample bam file. Total three multi-sample bam files, one for each regime, are generated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0dbe7f8-e7a9-4403-8829-41efa73164f7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output\n",
    "JOB_NAME=picard_MergeSamFiles\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=60G\n",
    "#SBATCH --time=0-48:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "regime_sample1=$2\n",
    "regime_sample2=$3\n",
    "regime_sample3=$4\n",
    "regime_sample4=$5\n",
    "regime_sample5=$6\n",
    "regime_sample6=$7\n",
    "\n",
    "module load java/jdk1.8.0_231\n",
    "module load picard/2.24.1\n",
    "\n",
    "\n",
    "java -jar /gpfs/software/ada/picard/2.24.1/picard.jar MergeSamFiles \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $regime_sample1)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $regime_sample2)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $regime_sample3)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $regime_sample4)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $regime_sample5)_viewed_marked_addg.bam \\\n",
    "I='$SCRATCH_DIRECTORY'/bam_files/$(basename $regime_sample6)_viewed_marked_addg.bam \\\n",
    "O='$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles.bam\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/MergeSamFiles.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_sample_wide | while read a b c d e f g; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/MergeSamFiles.sh \"$a\" \"$b\" \"$c\" \"$d\" \"$e\" \"$f\" \"$g\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/MergeSamFiles_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/MergeSamFiles_submit_jobs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24a252a7-8907-477f-ac34-40a47d804e82",
   "metadata": {},
   "source": [
    "<a id=\"46\"></a> \n",
    "## Add a readgroup for the ancestor to the header of the merged bam\n",
    "This script uses the samtools view tool to add a dummy readgroup for the ancestor to the merged, multi-sample bam file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4250d915-49f3-4d0a-8257-cbce5989e349",
   "metadata": {},
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/accuMUlate\n",
    "JOB_NAME=addg\n",
    "\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=12G\n",
    "#SBATCH --time=0-12:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "\n",
    "\n",
    "module load samtools/1.16.1\n",
    "\n",
    "samtools view -H '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles.bam > '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles.head.sam\n",
    "\n",
    "sed '\"'\"'19 i @RG\\tID:Ancestral\\tLB:lib1 PL:illumina_TruSeq\\tSM:Ancestral\\tPU:unit1'\"'\"' '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles.head.sam > '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_addg.head.sam\n",
    "\n",
    "samtools view '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles.bam | cat '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_addg.head.sam - | samtools view -Sb - > '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_addg.bam\n",
    "\n",
    "samtools index '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_addg.bam\n",
    "\n",
    "\n",
    "module load python/3.10.0\n",
    "\n",
    "samtools view -H '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_addg.bam | python '$HOME_DIRECTORY'/extract_samples.py Ancestral - >> '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_params.ini\n",
    "\n",
    "\n",
    "cat '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_params.ini '$HOME_DIRECTORY'/accuMUlate_parameters > '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_params_temp.ini && mv '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_params_temp.ini '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_params.ini\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/addg.sh\n",
    "\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_list | while read a; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/addg.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/addg_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/addg_submit_jobs\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eac87e24-1c01-4056-942c-6339144f3f42",
   "metadata": {},
   "source": [
    "<a id=\"47\"></a> \n",
    "## Call variants\n",
    "This script uses the accuMUlate tool to call variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db072559-c178-4982-85e4-22b332bdf170",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "REF_FASTA=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna.gz\n",
    "REF_FASTA_DECOMPRESSED=$SCRATCH_DIRECTORY/ref_fasta/GCA_001643735.4_ASM164373v4_genomic.fna\n",
    "JOB_OUTPUT=$SCRATCH_DIRECTORY/job_output/accuMUlate\n",
    "JOB_NAME=accuMUlate\n",
    "\n",
    "echo '#!/bin/bash -e\n",
    "#SBATCH --mail-type=ALL #Mail events (NONE, BEGIN, END, FAIL, ALL) \n",
    "#SBATCH --mail-user=papilio.chen@gmail.com #Where to send mail\n",
    "#SBATCH -p compute-24-96\n",
    "#SBATCH --mem=30G\n",
    "#SBATCH --time=5-00:00\n",
    "#SBATCH --job-name='$JOB_NAME'\n",
    "#SBATCH -o '$JOB_OUTPUT'/'$JOB_NAME'-%j.out\n",
    "#SBATCH -e '$JOB_OUTPUT'/'$JOB_NAME'-%j.err\n",
    "\n",
    "#ABOVE THIS LINE ARE THE SLURM DIRECTIVES\n",
    "#BELOW THIS LINE ARE THE COMMANDS TO RUN YOUR JOB\n",
    "\n",
    "regime=$1\n",
    "\n",
    "module load accuMUlate/0.2.2\n",
    "\n",
    "/gpfs/software/ada/accuMUlate/0.2.2/accuMUlate \\\n",
    "-b '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_addg.bam \\\n",
    "-r '$REF_FASTA_DECOMPRESSED' \\\n",
    "-q 13 \\\n",
    "-m 13 \\\n",
    "-p 0.10000000000000001 \\\n",
    "--ploidy-ancestor 2 \\\n",
    "--ploidy-descendant 2 \\\n",
    "-c '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_MergeSamFiles_params.ini \\\n",
    "-o '$HOME_DIRECTORY'/accuMUlate/$(basename $regime)_putative_mutations.tsv\n",
    "\n",
    "' > $HOME_DIRECTORY/scripts/accuMUlate.sh\n",
    "\n",
    "\n",
    "less $HOME_DIRECTORY/regime_list | while read a b c d e f g; do printf \"\n",
    "sbatch $HOME_DIRECTORY/scripts/accuMUlate.sh \"$a\"\n",
    "\"; done > $HOME_DIRECTORY/scripts/accuMUlate_submit_jobs\n",
    "\n",
    "# sh -e $HOME_DIRECTORY/scripts/accuMUlate_submit_jobs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04683015-f1db-46f3-a9b1-088c04547566",
   "metadata": {},
   "source": [
    "<a id=\"48\"></a> \n",
    "## Variant filtering\n",
    "This scripts filters to keep the following:\n",
    "1. Sites outside of soft-masked regions;\n",
    "2. Sites where estimated probability of at least one mutation at this site = 1;\n",
    "3. Sites where estimated probability of exactly one mutation at this site = 1; \n",
    "4. Sites where probability of the mutation direction above being correct = 1; \n",
    "5. Homozygous to heterozygous mutations;\n",
    "6. Sites where the number of reads for the mutant allele in non-mutant samples = 0; \n",
    "7. Sites where the allelic ratio for the mutant allele in the mutant sample is between 0.25 and 0.75;\n",
    "8. Sites where the AD test statistic for mapping quality difference between mutant and non-mutant alleles =< 1.95;\n",
    "9. Sites where the AD test statistic for insert size differfence between inferred insert size between mutant and non-mutant alleles =< 1.95;\n",
    "10. Sites where the p-value from Fisher's exact test of strand bias between mutant and non-mutant alleles is > 0.05;\n",
    "11. Sites where the p-value from Fisher's exact test of pair-mapping rate difference between mutant and non-mutant alleles is > 0.05."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73ede441-f4c2-457b-8460-afa08883cd11",
   "metadata": {},
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "less $HOME_DIRECTORY/regime_list | while read a; do printf \"\n",
    "\n",
    "awk '\"' $4~/[A-Z]/ '\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp1.tsv\n",
    "\n",
    "awk '\"' $7==1 && $8==1 && $9==1 {print $0}'\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp1.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp2.tsv\n",
    "\n",
    "awk '\"' { if (substr($6,1,1) == substr($6,2,1) && substr($6,1,1) == substr($6,5,1) && substr($6,5,1) != substr($6,6,1) ) print $0 }'\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp2.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp3.tsv\n",
    "\n",
    "awk '\"' $15==0 {print $0}'\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp3.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp4.tsv\n",
    "\n",
    "awk '\"' $14/($14+$12+$13) >=0.25 && $14/($14+$12+$13) <= 0.75 {print $0}'\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp4.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp5.tsv\n",
    "\n",
    "awk '\"' $16 <= 1.95 {print $0} '\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp5.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp6.tsv\n",
    "\n",
    "awk '\"' $17 <= 1.95 {print $0} '\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp6.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp7.tsv\n",
    "\n",
    "awk '\"' $18 > 0.05 {print $0} '\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp7.tsv > $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp8.tsv\n",
    "\n",
    "awk '\"' $19 > 0.05 {print $0} '\"' $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp8.tsv > $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach/\"$a\"_filtered_mutations.tsv\n",
    "\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp1.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp2.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp3.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp4.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp5.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp6.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp7.tsv\n",
    "rm $HOME_DIRECTORY/accuMUlate/\"$a\"_putative_mutations_temp8.tsv\n",
    "\"; done > $HOME_DIRECTORY/scripts/accuMUlate_filtering\n",
    "\n",
    "#sh -e $HOME_DIRECTORY/scripts/accuMUlate_filtering\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebf7d7dd-e3b2-475e-ac35-b59f31806cfe",
   "metadata": {},
   "source": [
    "<a id=\"49\"></a> \n",
    "## Derive the number of candidate germline mutations for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00451051-8568-4e28-b731-ffc4dd2e698b",
   "metadata": {},
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "touch $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach_number_GM\n",
    "\n",
    "less $HOME_DIRECTORY/regime_sample | while read a b; do printf \"\n",
    "grep '\"$b\"' $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach/\"$a\"_filtered_mutations.tsv > $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach/\"$b\"_filtered_mutations.tsv\n",
    "wc -l $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach/\"$b\"_filtered_mutations.tsv >> $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach_number_GM_before_manual_inspection\n",
    "\"; done > $HOME_DIRECTORY/scripts/accuMUlate_number_GM_before_manual_inspection\n",
    "\n",
    "sh -e $HOME_DIRECTORY/scripts/accuMUlate_number_GM_before_manual_inspection\n",
    "\n",
    "sed -i 's/\\/gpfs\\/home\\/wwh23wcu\\/de_novo_mutation\\/probabilistic_approach\\///' $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach_number_GM_before_manual_inspection\n",
    "\n",
    "sed -i 's/_filtered_mutations.tsv//' $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach_number_GM_before_manual_inspection\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06c27dff-56a8-4d17-a6ec-a572816c6901",
   "metadata": {},
   "source": [
    "<a id=\"51\"></a> \n",
    "## Derive bed files for manual inspection\n",
    "This script derives a bed file for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19ce92eb-640b-4c04-8550-9f1f8e31908a",
   "metadata": {},
   "outputs": [],
   "source": [
    "HOME_DIRECTORY=/gpfs/home/wwh23wcu\n",
    "SCRATCH_DIRECTORY=/gpfs/scratch/wwh23wcu\n",
    "\n",
    "less $HOME_DIRECTORY/regime_sample | while read a b; do echo \"\n",
    "\n",
    "awk '{print \\$1\\\" \\\"\\$3\\\" \\\"\\$4\\\" \\\"substr(\\$6,6,1)}' $HOME_DIRECTORY/de_novo_mutation/probabilistic_approach/\"$b\"_filtered_mutations.tsv > $HOME_DIRECTORY/de_novo_mutation/bed/\"$b\"_probabilistic_bed\n",
    "\n",
    "\"; done > $HOME_DIRECTORY/scripts/probabilistic_derive_bed\n",
    "\n",
    "sh -e $HOME_DIRECTORY/scripts/probabilistic_derive_bed\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "name": ""
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
